In [1]:
import os,sys,tqdm
import numpy as np
import tensorflow as tf
import tensorflow.keras
from tensorflow.keras.datasets import *
from tensorflow.keras.models import *
from tensorflow.keras.layers import *
from tensorflow.keras.preprocessing.text import *

from collections import Counter
import pandas as pd
import shutil
import pickle
import gc
# os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"
# The GPU id to use, usually either "0" or "1"
os.environ["CUDA_VISIBLE_DEVICES"]="2"

In [2]:
def basic_statistics(all_length):
    '''
    input: length list of elements e.g.[1,1,1,3,5,9,4,2,1,3,54,78,5...]
    output1: mean、std、mode、min、q1、median(q2)、q3、max、iqr、outlier、far out
    output2: statistics graph、10%~90% form
    '''
    stat_dict = {}
    stat_dict['mean'] = np.mean(all_length)
    stat_dict['std'] = np.std(all_length)
    stat_dict['mode'] = np.argmax(np.bincount(all_length))
    stat_dict['min'] = np.min(all_length)
    stat_dict['q1'] = np.quantile(all_length,0.25)
    stat_dict['median'] = np.quantile(all_length,0.5)
    stat_dict['q3'] = np.quantile(all_length,0.75)
    stat_dict['max'] = np.max(all_length)
    stat_dict['iqr'] = stat_dict['q3'] - stat_dict['q1']
    stat_dict['outlier'] = stat_dict['q3'] + 1.5*stat_dict['iqr']
    stat_dict['far_out'] = stat_dict['q3'] + 3*stat_dict['iqr']
    for i in [10,20,30,40,50,60,70,80,90,100]:
        stat_dict[str(i)+'%'] = np.percentile(all_length,i)
    return pd.DataFrame.from_dict(stat_dict,orient='index',columns=['length'])

In [2]:
max_words = 8352#8352 #Top most frequent words to consider. Any less frequent word will appear as oov_char value in the sequence data.
max_length = 360#360

In [3]:
word_index = reuters.get_word_index()
print('all_words#:',len(word_index))
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words,maxlen=max_length,
                                                         test_split=0.2,seed=830913)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')
word_index

all_words#: 30979
8260 train sequences
2066 test sequences


{'mdbl': 10996,
 'fawc': 16260,
 'degussa': 12089,
 'woods': 8803,
 'hanging': 13796,
 'localized': 20672,
 'sation': 20673,
 'chanthaburi': 20675,
 'refunding': 10997,
 'hermann': 8804,
 'passsengers': 20676,
 'stipulate': 20677,
 'heublein': 8352,
 'screaming': 20713,
 'tcby': 16261,
 'four': 185,
 'grains': 1642,
 'broiler': 20680,
 'wooden': 12090,
 'wednesday': 1220,
 'highveld': 13797,
 'duffour': 7593,
 '0053': 20681,
 'elections': 3914,
 '270': 2563,
 '271': 3551,
 '272': 5113,
 '273': 3552,
 '274': 3400,
 'rudman': 7975,
 '276': 3401,
 '277': 3478,
 '278': 3632,
 '279': 4309,
 'dormancy': 9381,
 'errors': 7247,
 'deferred': 3086,
 'sptnd': 20683,
 'cooking': 8805,
 'stratabit': 20684,
 'designing': 16262,
 'metalurgicos': 20685,
 'databank': 13798,
 '300er': 20686,
 'shocks': 20687,
 'nawg': 7972,
 'tnta': 20688,
 'perforations': 20689,
 'affiliates': 2891,
 '27p': 20690,
 'ching': 16263,
 'china': 595,
 'wagyu': 16264,
 'affiliated': 3189,
 'chino': 16265,
 'chinh': 16266,
 '

In [13]:
# train_len = [len(x) for x in x_train]
# test_len = [len(x) for x in x_test]
# all_len = train_len
# all_len.extend(test_len)
# basic_statistics(all_len)

length
mean      145.964197
std       145.878476
mode       17.000000
min         2.000000
q1         60.000000
median     95.000000
q3        180.000000
max      2376.000000
iqr       120.000000
outlier   360.000000
far_out   540.000000
10%        35.000000
20%        53.000000
30%        67.000000
40%        81.000000
50%        95.000000
60%       112.000000
70%       154.000000
80%       206.000000
90%       315.000000
100%     2376.000000

In [15]:
# df = pd.DataFrame(all_len)
# df.to_excel('./results/length_dist.xlsx', header=False, index=False)
# df

In [21]:
# train_words = []
# for x in x_train:
#     train_words.extend(x)
# test_words = []
# for x in x_test:
#     test_words.extend(x)
# all_words = train_words
# all_words.extend(test_words)
# all_statistcs = Counter(all_words)
# all_statistcs

Counter({1: 11228,
         53: 4213,
         352: 647,
         26: 8451,
         14: 15015,
         279: 801,
         39: 5818,
         72: 3091,
         4497: 26,
         18: 11039,
         83: 2597,
         5291: 21,
         88: 2381,
         5397: 20,
         11: 20141,
         3412: 37,
         19: 10755,
         151: 1363,
         230: 962,
         831: 253,
         15: 13329,
         165: 1232,
         318: 707,
         3780: 33,
         124: 1676,
         1527: 117,
         1424: 128,
         35: 6588,
         5302: 20,
         12: 16668,
         17: 11191,
         486: 459,
         341: 663,
         142: 1466,
         255: 870,
         219: 997,
         429: 528,
         68: 3363,
         146: 1402,
         252: 882,
         191: 1098,
         15448: 3,
         3631: 35,
         2283: 65,
         71: 3120,
         10: 29581,
         342: 660,
         49: 4565,
         1977: 80,
         324: 695,
         27: 8311,
         9222: 

In [28]:
# df = pd.DataFrame.from_dict(dict(all_statistcs), orient = 'index')
# df.to_excel('./results/words_dist2.xlsx', header=False, index=True)
# df

0
1      11228
53      4213
352      647
26      8451
14     15015
...      ...
24452      1
18567      2
27222      1
26864      1
24794      1

[30980 rows x 1 columns]

In [4]:
trainX = tf.keras.preprocessing.sequence.pad_sequences(x_train,maxlen=max_length,padding='post',value=0)
testX = tf.keras.preprocessing.sequence.pad_sequences(x_test,maxlen=max_length,padding='post',value=0)
print(trainX.shape, testX.shape)

(8260, 360) (2066, 360)


In [5]:
hidden_dim = 128
# do = 0.1

## Graph execution
### Embedder

In [42]:
int_id = Input(shape=(max_length,), dtype='int32', name='int_ids') # 輸入的api funvtion name ID
int_ids = Masking(mask_value=0)(int_id)
sent_emb = Embedding(max_words, hidden_dim,input_length=max_length
                    ,trainable=True,name='glove_emb')(int_ids) 

### Encoder

In [43]:
rnn = GRU(int(hidden_dim/2),return_sequences=True,return_state=False,name='common_extract'
                      ,trainable=True)(sent_emb)
rnn = BatchNormalization(name='bn')(rnn)


### Filter

In [44]:
fil = TimeDistributed(Dense(1,activation='sigmoid',
                             name='filter_out'),name='TD2')(rnn)

### Classfier

In [47]:
mul = Multiply()([fil,sent_emb])
clf = LSTM(int(hidden_dim/2),dropout=do,recurrent_dropout=do,name='lstm')(mul)
clf = BatchNormalization(name='bn3')(clf)
clf = Dense(max(y_train)+1,activation='softmax',name='clf')(clf)

## Compile

In [48]:
model = Model(inputs=int_id, outputs = clf)
model.summary()

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
int_ids (InputLayer)            [(None, 358)]        0                                            
__________________________________________________________________________________________________
masking_1 (Masking)             (None, 358)          0           int_ids[0][0]                    
__________________________________________________________________________________________________
glove_emb (Embedding)           (None, 358, 128)     1280000     masking_1[0][0]                  
__________________________________________________________________________________________________
common_extract (GRU)            (None, 358, 64)      37248       glove_emb[0][0]                  
______________________________________________________________________________________________

In [53]:
# loss
import keras.backend as K
def custom_objective(layer):
    return K.sum(layer.output)
#     return K.sum(layer.output)
# kk = tf.keras.backend.ea
model.compile(loss=custom_objective(model.get_layer(name='TD2')),optimizer='adam')

Using TensorFlow backend.


OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.

## Eager Execution

In [52]:
'''# whole model
do = 0
init = tensorflow.keras.initializers.Ones()
class base_model(Model):
    def __init__(self):
        super(base_model, self).__init__()
        self.mask = Masking(mask_value=0)
        self.emb = Embedding(max_words, hidden_dim,input_length=max_length
                    ,trainable=True,name='glove_emb')
        self.rnn1 = GRU(int(hidden_dim/2),return_sequences=True,return_state=False,name='common_extract'
                      ,trainable=True)
        self.bn1 = BatchNormalization(name='bn1')
        self.fil = Dense(1,activation='hard_sigmoid',kernel_initializer=init,bias_initializer=init,name='filter_out')
        #self.fil = TimeDistributed(Dense(1,activation='sigmoid', name='filter_out'),name='TD2')
        self.mul = Multiply()
        self.rnn2 = Bidirectional(GRU(int(hidden_dim/2),dropout=do,recurrent_dropout=do,name='lstm'))
        self.rnn3 = LSTM(int(hidden_dim/2))
        self.bn2 = BatchNormalization(name='bn2')
        self.out = Dense(max(y_train)+1,activation='softmax',name='clf')
    def transform(self,x):
        return tf.math.round(x)
    def call(self,x):
        x = self.mask(x)
        x1 = self.emb(x)
        x = self.rnn1(x1)
        x = self.bn1(x)
        y = self.fil(x)
        y1 = self.transform(y)
        x2 = self.mul([y1,x1])
        x = self.rnn2(x2) #x
        x = self.bn2(x)
        y2 = self.out(x)
        return y,y1,y2
        #return y,y1,y2,x2
        
model = base_model()
'''

In [277]:
# partial1 model
# init_w = tensorflow.keras.initializers.Constant(value=0.9) #portyion=0.6, w=0.9, b = 0.8-0.85 (0.83從0開始)
init_w = tensorflow.keras.initializers.RandomNormal(mean=1.667,stddev=0.2556)
# init_b = tensorflow.keras.initializers.Constant(value=0.84) #w=1 ; b=0.499, portion=1
init_b = tensorflow.keras.initializers.RandomNormal(mean=1.667,stddev=0.2756)
def onezero(x):
    portion = 0.6#0.6#0.6 #0.6~1
    z = tf.where(x>=1.0, x - x + 1.0, x)
    y = tf.where(z<=0.0, z - z + 0.0, portion*z)
    return y

class base_model_1(Model):
    def __init__(self):
        super(base_model_1, self).__init__()
        self.mask = Masking(mask_value=0)
        self.emb = Embedding(max_words, hidden_dim,input_length=max_length
                    ,trainable=True,name='glove_emb')
        self.rnn1 = GRU(int(hidden_dim/4),return_sequences=True,return_state=False,name='common_extract'
                      ,trainable=True)
        self.att = Attention(name='selfatt')
        self.bn1 = BatchNormalization(name='bn1')
        #self.fil = Dense(1,activation=onezero,name='filter_out')
        self.fil = TimeDistributed(Dense(1,activation=onezero,kernel_initializer=init_w,bias_initializer=init_b, name='filter_out'),name='TD2') #relu/linear/step function

    def call(self,x):
        x = self.mask(x)
        x1 = self.emb(x)
        x = self.att([x1,x1])
        #x = self.rnn1(x1)
        #x = self.bn1(x)
        y = self.fil(x)
        return x1,y

model1 = base_model_1()
#phase2
# model3 = load_model(saveP)

# model3 = base_model_1()
# model3.load_weights('./model/2019110501/model1')#,by_name=True)
# model1.emb.set_weights(model3.emb.get_weights())

In [278]:
# partial2 model
class base_model_2(Model):
    def __init__(self):
        super(base_model_2, self).__init__()
        self.mul = Multiply()
        #self.rnn2 = Bidirectional(GRU(int(hidden_dim/2),dropout=do,recurrent_dropout=do,name='lstm'))
        self.rnn3 = GRU(int(hidden_dim/2))
        self.bn2 = BatchNormalization(name='bn2')
        self.out = Dense(max(y_train)+1,activation='softmax',name='clf')

    def call(self,x1,y1):
        x2 = self.mul([y1,x1])
        x = self.rnn3(x2) #x2 #y1=weight|binary
        x = self.bn2(x)
        y2 = self.out(x)
        return y2
    
model2 = base_model_2()

In [448]:
'''
# TEST
x = tf.random.uniform((32, 6))
out1,out2 = model1(x)
out2 = tf.squeeze(out2,axis=-1)
# out2 = out2.astype('float32')
# out2 = tf.dtypes.cast(out2, tf.float8)
# print(out2.shape)
yy = tf.matmul(out2,kk)
print(yy[0].shape)
tf.where(yy[0]==0,1,0)

gg = tf.range(-5,5)
gg = tf.expand_dims(gg,axis=0)
gg = tf.keras.backend.repeat_elements(gg,rep=32,axis=0)
tf.where(gg==1,gg,0)

kk = tf.Variable(np.array([[1.0,1.0,1.0,0.0,0.0,0.0],[0.0,1.0,1.0,1.0,0.0,0.0],[0.0,0.0,1.0,1.0,1.0,0.0],
                           [0.0,0.0,0.0,1.0,1.0,1.0]]).T,dtype='float32')
kk = tf.expand_dims(kk,axis=0)
kk = tf.keras.backend.repeat_elements(kk,rep=1,axis=0)
kk.shape
'''

(32, 4)


<tf.Tensor: id=5770753, shape=(32, 4), dtype=int32, numpy=
array([[1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1]], dtype=int32)>

In [246]:
# 記得要跑到一個好的data
# batch_size = 128 #,reshuffle_each_iteration=True
# train_ds = tf.data.Dataset.from_tensor_slices((trainX,y_train)).shuffle(trainX.shape[0],reshuffle_each_iteration=True).batch(batch_size)
# valid_ds = tf.data.Dataset.from_tensor_slices((testX,y_test)).batch(batch_size)

In [279]:
seq_num = 2 #連續幾個才叫做連續，要改modify_idx看有幾個

arr_len = max_length - seq_num + 1
seq_arr = []
for i in range(arr_len):
    ori_np = np.array([0]*max_length)
    modify_idx = [i,i+1] #要跟著seq_num改
    ori_np[modify_idx]=1
    seq_arr.append(ori_np)
seq_arr = np.array(seq_arr)
seq_mask = tf.Variable(seq_arr.T,dtype='float32')
seq_mask = tf.expand_dims(seq_mask,axis=0)
seq_mask

<tf.Tensor: id=23103458, shape=(1, 360, 359), dtype=float32, numpy=
array([[[1., 0., 0., ..., 0., 0., 0.],
        [1., 1., 0., ..., 0., 0., 0.],
        [0., 1., 1., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 1., 1., 0.],
        [0., 0., 0., ..., 0., 1., 1.],
        [0., 0., 0., ..., 0., 0., 1.]]], dtype=float32)>

In [280]:
def loss_object1(predictions): #filter loss
    mask = tf.math.logical_not(tf.math.equal(predictions, 0))
    loss_ = tf.reduce_mean(predictions)
    mask = tf.cast(mask, dtype=loss_.dtype)
    loss_ *= mask
    return tf.reduce_mean(loss_)
def one_percentage(predictions): #1 num
    mask = tf.math.logical_not(tf.math.equal(predictions, 0))
    loss_ = tf.reduce_mean(predictions)
    mask = tf.cast(mask, dtype=loss_.dtype)
    loss_ *= mask
    return tf.reduce_mean(loss_)
def seq_loss(predictions):
    mask = tf.math.logical_not(tf.math.equal(predictions, 0))
    predictions = tf.squeeze(predictions,axis=-1)
    results = tf.matmul(predictions,seq_mask)
    results = tf.where(results==seq_num,1.0,0.0)
    loss_ = tf.reduce_mean(results)
    mask = tf.cast(mask, dtype=loss_.dtype)
    loss_ *= mask
    return tf.reduce_mean(loss_)
def word_reproduce1(prediction_int):
    return prediction_int
def word_reproduce2(prediction_int):
    return prediction_int
loss_object2 = tf.keras.losses.SparseCategoricalCrossentropy() #clf loss

optimizer1 = tf.keras.optimizers.SGD(learning_rate=0.004, momentum=0.9, nesterov=True)
# optimizer1 = tf.keras.optimizers.Nadam()#Adam(learning_rate=0.0003)
# optimizer2 = tf.keras.optimizers.RMSprop()

train_loss = tf.keras.metrics.Mean(name='train_loss') #total_loss
train_accloss = tf.keras.metrics.Mean(name='train_accloss')#loss_acc
train_filloss = tf.keras.metrics.Mean(name='train_filloss') #loss_filter
train_seqloss = tf.keras.metrics.Mean(name='train_seqloss') #loss_seq
train_ones = tf.keras.metrics.Mean(name='train_ones') #ones_num
train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='train_accuracy') #acc_rate

test_loss = tf.keras.metrics.Mean(name='test_loss') #total_loss
test_accloss = tf.keras.metrics.Mean(name='test_accloss')#loss_acc
test_filloss = tf.keras.metrics.Mean(name='test_filloss') #loss_filter
test_seqloss = tf.keras.metrics.Mean(name='test_seqloss')
test_ones = tf.keras.metrics.Mean(name='test_ones') #ones_num
test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='test_accuracy')

In [281]:
# seperate partial model
alpha = 0.000005#0.000002 #pahse1: -0.1 / 0.0 ; phase2: 0.01~0.05~0.1 有1-就是希望1越多，沒1-就是希望0越多1越少
beta = 1.0 #clf loss 越大越要求分好
gamma = 0.0 #seqloss 越大越要求連續
#-0.001 / 1.0 / 1.0

@tf.function
def train_step(x,yc):
    with tf.GradientTape(persistent=False) as tape:
        emb, pred_imp = model1(x)
        #loss1 = alpha*loss_object1(pred_imp) #phase1
        #pred_imp2 = tf.math.round(pred_imp)
        #pred_imp3 = tf.clip_by_value(pred_imp,clip_value_max=1,clip_value_min=0)
        pred_imp2 = tf.math.round(pred_imp)
        loss1 = loss_object1(pred_imp) #有1-就是希望1越多，沒1-就是希望0越多1越少 #pahse2:alpha*loss_object1(pred_imp) ; phase1: alpha*(1-loss_object1(pred_imp))
        pred_cat = model2(emb,pred_imp2) #weight: pred_imp; 0/1binary: pred_imp2 。#pahse1: pred_imp; phase2; pred_imp2
        loss2 = loss_object2(yc, pred_cat)
        loss3 = 1-seq_loss(pred_imp2)
        loss = alpha*loss1 + beta*loss2 + gamma*loss3
    trainable_variable = model1.trainable_variables
    trainable_variable.extend(model2.trainable_variables)
    gradients = tape.gradient(loss,trainable_variable)
    optimizer1.apply_gradients(zip(gradients,trainable_variable))
    
    train_loss(loss) #total_loss
    train_filloss(loss1)
    train_accloss(loss2)
    train_seqloss(loss3) #loss_seq
    train_accuracy(yc, pred_cat) #acc_rate
    ones = one_percentage(pred_imp2) #pred_imp2
    train_ones(ones) #ones_num
    
    
@tf.function
def test_step(x,yc):
    emb, pred_imp = model1(x)
    #loss1 = alpha*loss_object1(pred_imp) #phase1
    #pred_imp2 = tf.math.round(pred_imp)
    #pred_imp3 = tf.clip_by_value(pred_imp,clip_value_max=1,clip_value_min=0)
    pred_imp2 = tf.math.round(pred_imp)
    loss1 = loss_object1(pred_imp) #phase2
    pred_cat = model2(emb,pred_imp2) #weight: pred_imp ; binary:pred_imp2
    loss2 = loss_object2(yc, pred_cat)
    loss3 = 1-seq_loss(pred_imp2)
    #t_loss = loss1 + loss2
    t_loss = alpha*loss1 + beta*loss2 + gamma*loss3
    
    test_loss(t_loss)
    test_filloss(loss1)
    test_accloss(loss2)
    test_seqloss(loss3)
    test_accuracy(yc, pred_cat)
    t_ones = one_percentage(pred_imp2) #pred_imp2
    test_ones(t_ones)

    
word_idx_pd = pd.DataFrame(data=word_index,index=["ID"]).T
word_idx_pd['ori_word'] = word_idx_pd.index 
word_idx_pd = word_idx_pd.set_index(["ID"])
print(word_idx_pd)


# def getKeysByValues(listOfValues,dfOfElements=word_idx_pd):
#     ori_word_li = []
#     for item in listOfValues:
#         key = item.numpy()
#         if key == 0:
#             continue
#         ori_word = dfOfElements.loc[key][0]
#         ori_word_li.append(ori_word)
#     return ori_word_li
ep = -1
@tf.function
def test_step_inf(x,yc,epoch):
    #ori_text = getKeysByValues(list(x))
    emb, pred_imp = model1(x)
    #loss1 = alpha*loss_object1(pred_imp) #phase1
    #pred_imp2 = tf.math.round(pred_imp)
    #pred_imp3 = tf.clip_by_value(pred_imp,clip_value_max=1,clip_value_min=0)
    pred_imp2 = tf.math.round(pred_imp)
    #if ep != epoch:
        #ep = epoch
        #x = tf.dtypes.cast(x,dtype=tf.float32)
    pred_imp3 = tf.squeeze(pred_imp2)
    pred_imp3 = tf.dtypes.cast(pred_imp3,dtype=tf.int32)
    used_x = tf.math.multiply(x,pred_imp3)
        #word_reproduce1(x)
        #word_reproduce2(used_x)
        #used_text = getKeysByValues(list(used_x))
        #tf.print(used_x)
        #print(emb.numpy(),pred_imp, x,pred_imp3,used_x)
        #int_ori_text.append(x)
        #int_fil_text.append(used_x)
    
    loss1 = loss_object1(pred_imp) #phase2
    pred_cat = model2(emb,pred_imp2) #phase1: pred_imp ; phase2:pred_imp2
    loss2 = loss_object2(yc, pred_cat)
    loss3 = 1-seq_loss(pred_imp2)
    #t_loss = loss1 + loss2
    t_loss = alpha*loss1 + beta*loss2 + gamma*loss3
    
    test_loss(t_loss)
    test_filloss(loss1)
    test_accloss(loss2)
    test_seqloss(loss3)
    test_accuracy(yc, pred_cat)
    t_ones = one_percentage(pred_imp2) #pred_imp2
    test_ones(t_ones)
    return x, used_x

        ori_word
ID              
10996       mdbl
16260       fawc
12089    degussa
8803       woods
13796    hanging
...          ...
16258    rotting
10995       pods
2849       emery
30979  northerly
16259   onomichi

[30979 rows x 1 columns]


In [331]:
'''
#AIO
alpha = 0.1
beta = 1
gamma = 0
@tf.function
def train_step(x,yc):
    with tf.GradientTape(persistent=True) as tape: #persistent=True
        pred_imp,pred_round , pred_cat = model(x)
#         pred_cat = model(x)
#         loss = alpha*loss_object1(pred_imp) + loss_object2(yc,pred_cat)
        loss1 = alpha*loss_object1(pred_imp)
        loss2 = beta*loss_object2(yc,pred_cat)
        loss = loss_object2(yc,pred_cat)
#     gradients = tape.gradient(loss, model.trainable_variables)
    grad1 = tape.gradient(loss1, model.trainable_variables)
    grad2 = tape.gradient(loss2, model.trainable_variables)
#     optimizer1.apply_gradients(zip(gradients, model.trainable_variables))
    optimizer1.apply_gradients(zip(grad1, model.trainable_variables))
    optimizer2.apply_gradients(zip(grad2, model.trainable_variables))
#     with tf.GradientTape() as tape:
#         pred_imp , pred_cat = model(x)
#         loss2 = loss_object2(yc,pred_cat)
#         loss = alpha*loss_object1(pred_imp) + loss_object2(yc,pred_cat)
#     grad2 = tape.gradient(loss2, model.trainable_variables)
#     optimizer2.apply_gradients(zip(grad2, model.trainable_variables))

    train_loss(loss)
    train_accuracy(yc, pred_cat)
    ones = one_percentage(pred_round)
    train_ones(ones)
    
@tf.function
def test_step(x,yc):
    pred_imp,pred_round, pred_cat = model(x)
#     pred_cat = model(x)
    t_loss = alpha*loss_object1(pred_imp) + loss_object2(yc,pred_cat) 
#     t_loss = loss_object2(yc,pred_cat)
    
    test_loss(t_loss)
    test_accuracy(yc, pred_cat)
    t_ones = one_percentage(pred_round)
    test_ones(t_ones)
'''    

In [ ]:
EPOCHS = 2000
DateID = '2019112702'

saveP1 = './model/'+DateID+'/model1'
saveP2 = './model/'+DateID+'/model2'
train_loss_acc = []
train_loss_filter = []
train_loss_seq = []
train_weighted_loss = []
train_acc_rate = []
train_ones_num = []

test_loss_acc = []
test_loss_filter = []
test_loss_seq = []
test_weighted_loss = []
test_acc_rate = []
test_ones_num = []

int_ori_text =[]
int_fil_text = []
int_label = []
# signature_dict = {'att':model1.att}

gc.collect()
best_clf = 0.0
for epoch in range(EPOCHS):
    for text, labels in train_ds:
        train_step(text, labels)

    for test_text, test_labels in valid_ds:
        ori,fil = test_step_inf(test_text, test_labels,epoch) #with restore words
        #print(k2)
        #test_step(test_text, test_labels) #original
    int_ori_text.append(ori.numpy())
    int_fil_text.append(fil.numpy())
    int_label.append(test_labels.numpy())
    
    template = 'Epoch {}, Total Loss: {}, Clf Loss: {}, Filter Loss: {}, Seq Loss: {}, Accuracy Rate: {:5.2f}%, Ones Portion: {}, \
            Test_Total_Loss: {}, Test_Clf_Loss: {}, Test_Filter_Loss: {}, TEST_Seq_Loss: {}, Test_Accuracy_Rate: {:5.2f}%, Test_Ones_Portion: {}'
    print(template.format(epoch+1,train_loss.result(),
                          train_accloss.result(),train_filloss.result(),train_seqloss.result(),
                        train_accuracy.result()*100,train_ones.result(),
                        test_loss.result(),
                        test_accloss.result(),test_filloss.result(),test_seqloss.result(),
                        test_accuracy.result()*100,test_ones.result(),
                        ))
    #int_ori_text.append(word_reproduce1.result().numpy())
    #int_fil_text.append(word_reproduce2.result().numpy())
    train_loss_acc.append( train_accloss.result().numpy())
    train_loss_filter.append( train_filloss.result().numpy())
    train_loss_seq.append( train_seqloss.result().numpy())
    train_weighted_loss.append( train_loss.result().numpy())
    train_acc_rate.append( train_accuracy.result().numpy())
    train_ones_num.append( train_ones.result().numpy())
    
    test_loss_acc.append( test_accloss.result().numpy())
    test_loss_filter.append( test_filloss.result().numpy())
    test_loss_seq.append( test_seqloss.result().numpy())
    test_weighted_loss.append( test_loss.result().numpy())
    test_acc_rate.append( test_accuracy.result().numpy())
    test_ones_num.append( test_ones.result().numpy())
    if best_clf<=test_accuracy.result()*100:
        #tf.saved_model.save(model1,saveP1+'_all')
        #model1.save(saveP1,save_format='h5')
        #model2.save(saveP2,save_format='h5')
        #tf.saved_model.save(model2,saveP2+'_all')
        model1.save_weights(saveP1,save_format='tf')
        model2.save_weights(saveP2,save_format='tf')
        best_clf = test_accuracy.result()*100
        print('===MODEL WEIGHTS SAVED===',saveP1,saveP2)
    # Reset the metrics for the next epoch
    #word_reproduce1.reset_states()
    #word_reproduce2.reset_states()
    train_loss.reset_states()
    train_accloss.reset_states()
    train_filloss.reset_states()
    train_seqloss.reset_states()
    train_ones.reset_states()
    train_accuracy.reset_states()
    
    test_loss.reset_states()
    test_accloss.reset_states()
    test_filloss.reset_states()
    test_seqloss.reset_states()
    test_ones.reset_states()
    test_accuracy.reset_states()

Epoch 1, Total Loss: 3.099400758743286, Clf Loss: 3.099398374557495, Filter Loss: 0.5832521915435791, Seq Loss: 0.11047469079494476, Accuracy Rate: 35.04%, Ones Portion: 0.8896543979644775,             Test_Total_Loss: 2.6405112743377686, Test_Clf_Loss: 2.6405086517333984, Test_Filter_Loss: 0.52372145652771, TEST_Seq_Loss: 0.661488950252533, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.33921027183532715
===MODEL WEIGHTS SAVED=== ./model/2019112702/model1 ./model/2019112702/model2
Epoch 2, Total Loss: 2.4706435203552246, Clf Loss: 2.4706408977508545, Filter Loss: 0.5493626594543457, Seq Loss: 0.3163779675960541, Accuracy Rate: 37.72%, Ones Portion: 0.6839318871498108,             Test_Total_Loss: 2.4235663414001465, Test_Clf_Loss: 2.4235637187957764, Test_Filter_Loss: 0.5501726269721985, TEST_Seq_Loss: 0.3087596893310547, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.6915099024772644
===MODEL WEIGHTS SAVED=== ./model/2019112702/model1 ./model/2019112702/model2
Epoch 3, Total Loss

Epoch 20, Total Loss: 2.270428419113159, Clf Loss: 2.2704250812530518, Filter Loss: 0.5306491255760193, Seq Loss: 0.5876320004463196, Accuracy Rate: 37.51%, Ones Portion: 0.4129648208618164,             Test_Total_Loss: 2.27176570892334, Test_Clf_Loss: 2.271763324737549, Test_Filter_Loss: 0.5299664735794067, TEST_Seq_Loss: 0.6176204681396484, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.3829679787158966
Epoch 21, Total Loss: 2.26098895072937, Clf Loss: 2.260986089706421, Filter Loss: 0.5300360321998596, Seq Loss: 0.6052649021148682, Accuracy Rate: 37.74%, Ones Portion: 0.3953196406364441,             Test_Total_Loss: 2.28365421295166, Test_Clf_Loss: 2.283651351928711, Test_Filter_Loss: 0.5360197424888611, TEST_Seq_Loss: 0.5234215259552002, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.4770996570587158
Epoch 22, Total Loss: 2.264296531677246, Clf Loss: 2.2642929553985596, Filter Loss: 0.531596839427948, Seq Loss: 0.585905134677887, Accuracy Rate: 37.46%, Ones Portion: 0.414647191

Epoch 40, Total Loss: 2.279104709625244, Clf Loss: 2.279102325439453, Filter Loss: 0.5188926458358765, Seq Loss: 0.7421483397483826, Accuracy Rate: 37.47%, Ones Portion: 0.25852543115615845,             Test_Total_Loss: 2.278980255126953, Test_Clf_Loss: 2.278977870941162, Test_Filter_Loss: 0.5231168866157532, TEST_Seq_Loss: 0.7099225521087646, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.29082930088043213
Epoch 41, Total Loss: 2.2751805782318115, Clf Loss: 2.275177240371704, Filter Loss: 0.5184797644615173, Seq Loss: 0.7427811622619629, Accuracy Rate: 37.71%, Ones Portion: 0.25790128111839294,             Test_Total_Loss: 2.2826271057128906, Test_Clf_Loss: 2.2826242446899414, Test_Filter_Loss: 0.5185315608978271, TEST_Seq_Loss: 0.7520356178283691, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.24864813685417175
Epoch 42, Total Loss: 2.278149127960205, Clf Loss: 2.278146982192993, Filter Loss: 0.5179998874664307, Seq Loss: 0.7454262375831604, Accuracy Rate: 37.52%, Ones Portion: 0

Epoch 60, Total Loss: 2.287653923034668, Clf Loss: 2.287651777267456, Filter Loss: 0.5119679570198059, Seq Loss: 0.7956172227859497, Accuracy Rate: 37.75%, Ones Portion: 0.20512834191322327,             Test_Total_Loss: 2.2871153354644775, Test_Clf_Loss: 2.2871124744415283, Test_Filter_Loss: 0.5114724636077881, TEST_Seq_Loss: 0.807007372379303, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.19366595149040222
Epoch 61, Total Loss: 2.2863030433654785, Clf Loss: 2.2863004207611084, Filter Loss: 0.5128798484802246, Seq Loss: 0.7916156053543091, Accuracy Rate: 37.57%, Ones Portion: 0.2090769112110138,             Test_Total_Loss: 2.284849166870117, Test_Clf_Loss: 2.284846305847168, Test_Filter_Loss: 0.5142558217048645, TEST_Seq_Loss: 0.7887012958526611, Test_Accuracy_Rate: 37.17%, Test_Ones_Portion: 0.21206901967525482
Epoch 62, Total Loss: 2.280674457550049, Clf Loss: 2.2806713581085205, Filter Loss: 0.5137882828712463, Seq Loss: 0.7876949906349182, Accuracy Rate: 37.85%, Ones Portion: 0

Epoch 80, Total Loss: 2.291175603866577, Clf Loss: 2.291172981262207, Filter Loss: 0.4976024925708771, Seq Loss: 0.8488046526908875, Accuracy Rate: 37.69%, Ones Portion: 0.15185202658176422,             Test_Total_Loss: 2.2928431034088135, Test_Clf_Loss: 2.2928409576416016, Test_Filter_Loss: 0.4981556534767151, TEST_Seq_Loss: 0.851800262928009, Test_Accuracy_Rate: 38.96%, Test_Ones_Portion: 0.1489088535308838
===MODEL WEIGHTS SAVED=== ./model/2019112702/model1 ./model/2019112702/model2
Epoch 81, Total Loss: 2.291290044784546, Clf Loss: 2.291287422180176, Filter Loss: 0.5004313588142395, Seq Loss: 0.8429602384567261, Accuracy Rate: 37.97%, Ones Portion: 0.15772387385368347,             Test_Total_Loss: 2.293560743331909, Test_Clf_Loss: 2.293558120727539, Test_Filter_Loss: 0.4964826703071594, TEST_Seq_Loss: 0.8542167544364929, Test_Accuracy_Rate: 39.01%, Test_Ones_Portion: 0.1464722454547882
===MODEL WEIGHTS SAVED=== ./model/2019112702/model1 ./model/2019112702/model2
Epoch 82, Total Los

Epoch 99, Total Loss: 2.2925355434417725, Clf Loss: 2.2925338745117188, Filter Loss: 0.4949226379394531, Seq Loss: 0.857415497303009, Accuracy Rate: 37.97%, Ones Portion: 0.14328326284885406,             Test_Total_Loss: 2.2944564819335938, Test_Clf_Loss: 2.294454574584961, Test_Filter_Loss: 0.495261013507843, TEST_Seq_Loss: 0.858449399471283, Test_Accuracy_Rate: 38.92%, Test_Ones_Portion: 0.14226585626602173
Epoch 100, Total Loss: 2.2916293144226074, Clf Loss: 2.2916274070739746, Filter Loss: 0.4954756498336792, Seq Loss: 0.8566497564315796, Accuracy Rate: 38.05%, Ones Portion: 0.14403671026229858,             Test_Total_Loss: 2.2949914932250977, Test_Clf_Loss: 2.2949891090393066, Test_Filter_Loss: 0.49372971057891846, TEST_Seq_Loss: 0.8609585762023926, Test_Accuracy_Rate: 38.67%, Test_Ones_Portion: 0.13975445926189423
Epoch 101, Total Loss: 2.2914657592773438, Clf Loss: 2.2914631366729736, Filter Loss: 0.49427923560142517, Seq Loss: 0.8589918613433838, Accuracy Rate: 38.00%, Ones Por

Epoch 119, Total Loss: 2.2954845428466797, Clf Loss: 2.295482635498047, Filter Loss: 0.4791211187839508, Seq Loss: 0.8872466087341309, Accuracy Rate: 38.24%, Ones Portion: 0.11349574476480484,             Test_Total_Loss: 2.2924387454986572, Test_Clf_Loss: 2.2924365997314453, Test_Filter_Loss: 0.480215460062027, TEST_Seq_Loss: 0.8834745287895203, Test_Accuracy_Rate: 38.63%, Test_Ones_Portion: 0.11729130893945694
Epoch 120, Total Loss: 2.2941763401031494, Clf Loss: 2.2941734790802, Filter Loss: 0.48132917284965515, Seq Loss: 0.8832908868789673, Accuracy Rate: 38.14%, Ones Portion: 0.1174536943435669,             Test_Total_Loss: 2.291827440261841, Test_Clf_Loss: 2.2918248176574707, Test_Filter_Loss: 0.48081958293914795, TEST_Seq_Loss: 0.8829160332679749, Test_Accuracy_Rate: 38.63%, Test_Ones_Portion: 0.11784316599369049
Epoch 121, Total Loss: 2.291950225830078, Clf Loss: 2.291947603225708, Filter Loss: 0.4798373579978943, Seq Loss: 0.886822521686554, Accuracy Rate: 38.34%, Ones Portion:

Epoch 139, Total Loss: 2.3003652095794678, Clf Loss: 2.3003623485565186, Filter Loss: 0.4710915684700012, Seq Loss: 0.9006518125534058, Accuracy Rate: 38.28%, Ones Portion: 0.10016307979822159,             Test_Total_Loss: 2.2984821796417236, Test_Clf_Loss: 2.2984797954559326, Test_Filter_Loss: 0.469541996717453, TEST_Seq_Loss: 0.8992202877998352, Test_Accuracy_Rate: 38.19%, Test_Ones_Portion: 0.1014958843588829
Epoch 140, Total Loss: 2.302541494369507, Clf Loss: 2.3025388717651367, Filter Loss: 0.4684560298919678, Seq Loss: 0.9041048288345337, Accuracy Rate: 38.22%, Ones Portion: 0.09671871364116669,             Test_Total_Loss: 2.299734115600586, Test_Clf_Loss: 2.299731731414795, Test_Filter_Loss: 0.46776193380355835, TEST_Seq_Loss: 0.9019569754600525, Test_Accuracy_Rate: 38.24%, Test_Ones_Portion: 0.09875510632991791
Epoch 141, Total Loss: 2.2989399433135986, Clf Loss: 2.2989370822906494, Filter Loss: 0.46720921993255615, Seq Loss: 0.9064642786979675, Accuracy Rate: 38.22%, Ones Por

Epoch 159, Total Loss: 2.2997124195098877, Clf Loss: 2.2997100353240967, Filter Loss: 0.47288408875465393, Seq Loss: 0.8987772464752197, Accuracy Rate: 38.28%, Ones Portion: 0.1020151749253273,             Test_Total_Loss: 2.3013381958007812, Test_Clf_Loss: 2.3013360500335693, Test_Filter_Loss: 0.46865057945251465, TEST_Seq_Loss: 0.9016022086143494, Test_Accuracy_Rate: 38.24%, Test_Ones_Portion: 0.09911813586950302
Epoch 160, Total Loss: 2.302353620529175, Clf Loss: 2.302351236343384, Filter Loss: 0.4680085778236389, Seq Loss: 0.9056248068809509, Accuracy Rate: 38.26%, Ones Portion: 0.09520606696605682,             Test_Total_Loss: 2.299955129623413, Test_Clf_Loss: 2.299952507019043, Test_Filter_Loss: 0.46981921792030334, TEST_Seq_Loss: 0.9000318646430969, Test_Accuracy_Rate: 38.24%, Test_Ones_Portion: 0.1007147803902626
Epoch 161, Total Loss: 2.300626039505005, Clf Loss: 2.3006231784820557, Filter Loss: 0.4727208912372589, Seq Loss: 0.8988566398620605, Accuracy Rate: 38.27%, Ones Port

Epoch 179, Total Loss: 2.300471782684326, Clf Loss: 2.3004701137542725, Filter Loss: 0.47154760360717773, Seq Loss: 0.9011020660400391, Accuracy Rate: 38.18%, Ones Portion: 0.0996968224644661,             Test_Total_Loss: 2.2990527153015137, Test_Clf_Loss: 2.2990503311157227, Test_Filter_Loss: 0.4721169173717499, TEST_Seq_Loss: 0.8973488807678223, Test_Accuracy_Rate: 38.24%, Test_Ones_Portion: 0.10343955457210541
Epoch 180, Total Loss: 2.297410249710083, Clf Loss: 2.297407627105713, Filter Loss: 0.4727678596973419, Seq Loss: 0.9000541567802429, Accuracy Rate: 38.22%, Ones Portion: 0.10072938352823257,             Test_Total_Loss: 2.3023035526275635, Test_Clf_Loss: 2.3023009300231934, Test_Filter_Loss: 0.4707031846046448, TEST_Seq_Loss: 0.8998962044715881, Test_Accuracy_Rate: 38.33%, Test_Ones_Portion: 0.10090240091085434
Epoch 181, Total Loss: 2.296569347381592, Clf Loss: 2.296566963195801, Filter Loss: 0.4725218415260315, Seq Loss: 0.9000735878944397, Accuracy Rate: 38.21%, Ones Porti

Epoch 199, Total Loss: 2.2991623878479004, Clf Loss: 2.2991597652435303, Filter Loss: 0.47274142503738403, Seq Loss: 0.9003096222877502, Accuracy Rate: 38.24%, Ones Portion: 0.10047803819179535,             Test_Total_Loss: 2.3023509979248047, Test_Clf_Loss: 2.3023483753204346, Test_Filter_Loss: 0.4715583324432373, TEST_Seq_Loss: 0.8998470902442932, Test_Accuracy_Rate: 38.24%, Test_Ones_Portion: 0.10095518827438354
Epoch 200, Total Loss: 2.298251152038574, Clf Loss: 2.2982497215270996, Filter Loss: 0.46824485063552856, Seq Loss: 0.9069128036499023, Accuracy Rate: 38.21%, Ones Portion: 0.09389778971672058,             Test_Total_Loss: 2.304248809814453, Test_Clf_Loss: 2.304246664047241, Test_Filter_Loss: 0.46900856494903564, TEST_Seq_Loss: 0.9036084413528442, Test_Accuracy_Rate: 38.00%, Test_Ones_Portion: 0.0971565917134285
Epoch 201, Total Loss: 2.2980504035949707, Clf Loss: 2.298048496246338, Filter Loss: 0.4710284173488617, Seq Loss: 0.903631865978241, Accuracy Rate: 38.23%, Ones Por

Epoch 219, Total Loss: 2.2989308834075928, Clf Loss: 2.2989284992218018, Filter Loss: 0.47094452381134033, Seq Loss: 0.9041807055473328, Accuracy Rate: 38.20%, Ones Portion: 0.09663759171962738,             Test_Total_Loss: 2.304537057876587, Test_Clf_Loss: 2.304534435272217, Test_Filter_Loss: 0.4695509374141693, TEST_Seq_Loss: 0.9038072228431702, Test_Accuracy_Rate: 37.95%, Test_Ones_Portion: 0.09698998928070068
Epoch 220, Total Loss: 2.299009084701538, Clf Loss: 2.299006938934326, Filter Loss: 0.47346699237823486, Seq Loss: 0.9008063077926636, Accuracy Rate: 38.26%, Ones Portion: 0.099970743060112,             Test_Total_Loss: 2.304619073867798, Test_Clf_Loss: 2.304616928100586, Test_Filter_Loss: 0.46955129504203796, TEST_Seq_Loss: 0.9039788842201233, Test_Accuracy_Rate: 37.90%, Test_Ones_Portion: 0.09682173281908035
Epoch 221, Total Loss: 2.299638032913208, Clf Loss: 2.299635410308838, Filter Loss: 0.47101104259490967, Seq Loss: 0.9041599631309509, Accuracy Rate: 38.23%, Ones Portio

Epoch 239, Total Loss: 2.3044235706329346, Clf Loss: 2.304421901702881, Filter Loss: 0.46871647238731384, Seq Loss: 0.9072553515434265, Accuracy Rate: 38.11%, Ones Portion: 0.0935555249452591,             Test_Total_Loss: 2.3034303188323975, Test_Clf_Loss: 2.3034274578094482, Test_Filter_Loss: 0.47146663069725037, TEST_Seq_Loss: 0.9013738632202148, Test_Accuracy_Rate: 38.00%, Test_Ones_Portion: 0.09941363334655762
Epoch 240, Total Loss: 2.29856276512146, Clf Loss: 2.298560619354248, Filter Loss: 0.4716520607471466, Seq Loss: 0.90412837266922, Accuracy Rate: 38.15%, Ones Portion: 0.09664808213710785,             Test_Total_Loss: 2.3046047687530518, Test_Clf_Loss: 2.3046019077301025, Test_Filter_Loss: 0.4709126651287079, TEST_Seq_Loss: 0.9028058648109436, Test_Accuracy_Rate: 37.90%, Test_Ones_Portion: 0.09799641370773315
Epoch 241, Total Loss: 2.3010025024414062, Clf Loss: 2.300999879837036, Filter Loss: 0.47168248891830444, Seq Loss: 0.9035362005233765, Accuracy Rate: 38.24%, Ones Porti

Epoch 259, Total Loss: 2.3065130710601807, Clf Loss: 2.3065102100372314, Filter Loss: 0.4697408676147461, Seq Loss: 0.9067586064338684, Accuracy Rate: 38.18%, Ones Portion: 0.09405705332756042,             Test_Total_Loss: 2.3114538192749023, Test_Clf_Loss: 2.3114516735076904, Test_Filter_Loss: 0.4681456983089447, TEST_Seq_Loss: 0.9078812599182129, Test_Accuracy_Rate: 37.85%, Test_Ones_Portion: 0.09292420744895935
Epoch 260, Total Loss: 2.2997958660125732, Clf Loss: 2.299793004989624, Filter Loss: 0.46933668851852417, Seq Loss: 0.9073312878608704, Accuracy Rate: 38.06%, Ones Portion: 0.09346981346607208,             Test_Total_Loss: 2.3104655742645264, Test_Clf_Loss: 2.3104634284973145, Test_Filter_Loss: 0.4684056341648102, TEST_Seq_Loss: 0.9076545834541321, Test_Accuracy_Rate: 37.85%, Test_Ones_Portion: 0.09314851462841034
Epoch 261, Total Loss: 2.3042404651641846, Clf Loss: 2.3042378425598145, Filter Loss: 0.4682236611843109, Seq Loss: 0.9090411067008972, Accuracy Rate: 38.06%, Ones 

Epoch 279, Total Loss: 2.3021180629730225, Clf Loss: 2.3021161556243896, Filter Loss: 0.46936509013175964, Seq Loss: 0.9079732298851013, Accuracy Rate: 38.00%, Ones Portion: 0.09283459931612015,             Test_Total_Loss: 2.310508966445923, Test_Clf_Loss: 2.310506582260132, Test_Filter_Loss: 0.4688059687614441, TEST_Seq_Loss: 0.9083694219589233, Test_Accuracy_Rate: 37.66%, Test_Ones_Portion: 0.09244298934936523
Epoch 280, Total Loss: 2.3011646270751953, Clf Loss: 2.301162004470825, Filter Loss: 0.4706052243709564, Seq Loss: 0.9060630202293396, Accuracy Rate: 38.06%, Ones Portion: 0.09473192691802979,             Test_Total_Loss: 2.310568332672119, Test_Clf_Loss: 2.3105664253234863, Test_Filter_Loss: 0.4708433151245117, TEST_Seq_Loss: 0.9049690365791321, Test_Accuracy_Rate: 37.90%, Test_Ones_Portion: 0.09580940753221512
Epoch 281, Total Loss: 2.3050408363342285, Clf Loss: 2.3050386905670166, Filter Loss: 0.46538829803466797, Seq Loss: 0.9126973152160645, Accuracy Rate: 37.97%, Ones Po

Epoch 299, Total Loss: 2.313699722290039, Clf Loss: 2.3136978149414062, Filter Loss: 0.4447380602359772, Seq Loss: 0.9344574809074402, Accuracy Rate: 38.21%, Ones Portion: 0.06633112579584122,             Test_Total_Loss: 2.3230786323547363, Test_Clf_Loss: 2.3230762481689453, Test_Filter_Loss: 0.4447672963142395, TEST_Seq_Loss: 0.9314649105072021, Test_Accuracy_Rate: 38.09%, Test_Ones_Portion: 0.06934653967618942
Epoch 300, Total Loss: 2.3116424083709717, Clf Loss: 2.3116400241851807, Filter Loss: 0.44377601146698, Seq Loss: 0.9344236254692078, Accuracy Rate: 38.15%, Ones Portion: 0.06635826826095581,             Test_Total_Loss: 2.323110580444336, Test_Clf_Loss: 2.323108434677124, Test_Filter_Loss: 0.4434739649295807, TEST_Seq_Loss: 0.9324702024459839, Test_Accuracy_Rate: 38.14%, Test_Ones_Portion: 0.06832510232925415
Epoch 301, Total Loss: 2.3118388652801514, Clf Loss: 2.3118367195129395, Filter Loss: 0.4434793293476105, Seq Loss: 0.9352203607559204, Accuracy Rate: 38.18%, Ones Porti

Epoch 319, Total Loss: 2.3139891624450684, Clf Loss: 2.3139877319335938, Filter Loss: 0.42770126461982727, Seq Loss: 0.9459806084632874, Accuracy Rate: 38.20%, Ones Portion: 0.05481494963169098,             Test_Total_Loss: 2.329209566116333, Test_Clf_Loss: 2.329206943511963, Test_Filter_Loss: 0.42655202746391296, TEST_Seq_Loss: 0.9443084597587585, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.05632099509239197
Epoch 320, Total Loss: 2.315978765487671, Clf Loss: 2.3159754276275635, Filter Loss: 0.4261184632778168, Seq Loss: 0.9470813274383545, Accuracy Rate: 38.16%, Ones Portion: 0.05372341349720955,             Test_Total_Loss: 2.330496311187744, Test_Clf_Loss: 2.3304946422576904, Test_Filter_Loss: 0.42659619450569153, TEST_Seq_Loss: 0.9442641139030457, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.056371044367551804
Epoch 321, Total Loss: 2.3165721893310547, Clf Loss: 2.3165698051452637, Filter Loss: 0.426288366317749, Seq Loss: 0.9464431405067444, Accuracy Rate: 38.16%, Ones P

Epoch 339, Total Loss: 2.3142647743225098, Clf Loss: 2.314262866973877, Filter Loss: 0.41839370131492615, Seq Loss: 0.9514082074165344, Accuracy Rate: 38.04%, Ones Portion: 0.04932880401611328,             Test_Total_Loss: 2.33280086517334, Test_Clf_Loss: 2.332798719406128, Test_Filter_Loss: 0.41853007674217224, TEST_Seq_Loss: 0.9481514096260071, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.05250288173556328
Epoch 340, Total Loss: 2.315941572189331, Clf Loss: 2.3159396648406982, Filter Loss: 0.4174725115299225, Seq Loss: 0.9515717625617981, Accuracy Rate: 38.04%, Ones Portion: 0.049144815653562546,             Test_Total_Loss: 2.332366943359375, Test_Clf_Loss: 2.332365036010742, Test_Filter_Loss: 0.4181886613368988, TEST_Seq_Loss: 0.9482802152633667, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.05237220972776413
Epoch 341, Total Loss: 2.319262981414795, Clf Loss: 2.3192601203918457, Filter Loss: 0.4187842011451721, Seq Loss: 0.950928270816803, Accuracy Rate: 38.03%, Ones Portio

Epoch 359, Total Loss: 2.3215489387512207, Clf Loss: 2.321546792984009, Filter Loss: 0.3940923810005188, Seq Loss: 0.9607862234115601, Accuracy Rate: 37.89%, Ones Portion: 0.03973471000790596,             Test_Total_Loss: 2.336111307144165, Test_Clf_Loss: 2.3361093997955322, Test_Filter_Loss: 0.39462441205978394, TEST_Seq_Loss: 0.9585137963294983, Test_Accuracy_Rate: 37.85%, Test_Ones_Portion: 0.041945282369852066
Epoch 360, Total Loss: 2.321143865585327, Clf Loss: 2.321141481399536, Filter Loss: 0.3916299641132355, Seq Loss: 0.9617867469787598, Accuracy Rate: 37.92%, Ones Portion: 0.03871743753552437,             Test_Total_Loss: 2.3351287841796875, Test_Clf_Loss: 2.3351268768310547, Test_Filter_Loss: 0.39191189408302307, TEST_Seq_Loss: 0.9591507911682129, Test_Accuracy_Rate: 37.90%, Test_Ones_Portion: 0.04127831012010574
Epoch 361, Total Loss: 2.3191916942596436, Clf Loss: 2.31919002532959, Filter Loss: 0.38980555534362793, Seq Loss: 0.9623246788978577, Accuracy Rate: 37.92%, Ones Po

Epoch 379, Total Loss: 2.323951005935669, Clf Loss: 2.323949098587036, Filter Loss: 0.36670204997062683, Seq Loss: 0.9674370884895325, Accuracy Rate: 37.89%, Ones Portion: 0.03291647508740425,             Test_Total_Loss: 2.3379030227661133, Test_Clf_Loss: 2.3379008769989014, Test_Filter_Loss: 0.36761608719825745, TEST_Seq_Loss: 0.9647586941719055, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.03562929108738899
Epoch 380, Total Loss: 2.3225364685058594, Clf Loss: 2.3225345611572266, Filter Loss: 0.3651885986328125, Seq Loss: 0.9677938222885132, Accuracy Rate: 37.88%, Ones Portion: 0.032550424337387085,             Test_Total_Loss: 2.3379759788513184, Test_Clf_Loss: 2.3379740715026855, Test_Filter_Loss: 0.3669525980949402, TEST_Seq_Loss: 0.9648681879043579, Test_Accuracy_Rate: 37.80%, Test_Ones_Portion: 0.03551296889781952
Epoch 381, Total Loss: 2.3264522552490234, Clf Loss: 2.3264503479003906, Filter Loss: 0.3641023635864258, Seq Loss: 0.9675503969192505, Accuracy Rate: 37.88%, Ones

Epoch 399, Total Loss: 2.322859048843384, Clf Loss: 2.322856903076172, Filter Loss: 0.32492223381996155, Seq Loss: 0.9732785224914551, Accuracy Rate: 37.89%, Ones Portion: 0.02702924609184265,             Test_Total_Loss: 2.343235731124878, Test_Clf_Loss: 2.343234062194824, Test_Filter_Loss: 0.3264630138874054, TEST_Seq_Loss: 0.969918966293335, Test_Accuracy_Rate: 37.66%, Test_Ones_Portion: 0.030292026698589325
Epoch 400, Total Loss: 2.3263137340545654, Clf Loss: 2.3263118267059326, Filter Loss: 0.32366007566452026, Seq Loss: 0.9732210040092468, Accuracy Rate: 37.89%, Ones Portion: 0.027076683938503265,             Test_Total_Loss: 2.3426294326782227, Test_Clf_Loss: 2.34262752532959, Test_Filter_Loss: 0.3250277638435364, TEST_Seq_Loss: 0.9700102210044861, Test_Accuracy_Rate: 37.66%, Test_Ones_Portion: 0.030196893960237503
Epoch 401, Total Loss: 2.3245325088500977, Clf Loss: 2.324530601501465, Filter Loss: 0.32143187522888184, Seq Loss: 0.9735113978385925, Accuracy Rate: 37.88%, Ones Po

Epoch 419, Total Loss: 2.3279693126678467, Clf Loss: 2.327968120574951, Filter Loss: 0.2812950313091278, Seq Loss: 0.9775763154029846, Accuracy Rate: 37.89%, Ones Portion: 0.022674402222037315,             Test_Total_Loss: 2.352867364883423, Test_Clf_Loss: 2.3528659343719482, Test_Filter_Loss: 0.28313887119293213, TEST_Seq_Loss: 0.9736666083335876, Test_Accuracy_Rate: 37.66%, Test_Ones_Portion: 0.02659159153699875
Epoch 420, Total Loss: 2.32840633392334, Clf Loss: 2.3284051418304443, Filter Loss: 0.2798715829849243, Seq Loss: 0.9776244163513184, Accuracy Rate: 37.92%, Ones Portion: 0.02262825146317482,             Test_Total_Loss: 2.353574275970459, Test_Clf_Loss: 2.3535726070404053, Test_Filter_Loss: 0.28034865856170654, TEST_Seq_Loss: 0.9739086031913757, Test_Accuracy_Rate: 37.66%, Test_Ones_Portion: 0.026364833116531372
Epoch 421, Total Loss: 2.3262152671813965, Clf Loss: 2.3262133598327637, Filter Loss: 0.27748265862464905, Seq Loss: 0.9779438376426697, Accuracy Rate: 37.91%, Ones 

Epoch 439, Total Loss: 2.3325159549713135, Clf Loss: 2.332515001296997, Filter Loss: 0.23346103727817535, Seq Loss: 0.9816323518753052, Accuracy Rate: 37.91%, Ones Portion: 0.01870674081146717,             Test_Total_Loss: 2.3538050651550293, Test_Clf_Loss: 2.353804111480713, Test_Filter_Loss: 0.2351093292236328, TEST_Seq_Loss: 0.9779167175292969, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.02237308956682682
Epoch 440, Total Loss: 2.330054521560669, Clf Loss: 2.3300533294677734, Filter Loss: 0.23068784177303314, Seq Loss: 0.9819527268409729, Accuracy Rate: 37.91%, Ones Portion: 0.018385084345936775,             Test_Total_Loss: 2.354830265045166, Test_Clf_Loss: 2.3548293113708496, Test_Filter_Loss: 0.23302781581878662, TEST_Seq_Loss: 0.9781020879745483, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.022185591980814934
Epoch 441, Total Loss: 2.3310208320617676, Clf Loss: 2.331019639968872, Filter Loss: 0.22820794582366943, Seq Loss: 0.9823693037033081, Accuracy Rate: 37.91%, Ones

Epoch 459, Total Loss: 2.331014633178711, Clf Loss: 2.3310134410858154, Filter Loss: 0.20356464385986328, Seq Loss: 0.9837063550949097, Accuracy Rate: 37.87%, Ones Portion: 0.01655367575585842,             Test_Total_Loss: 2.3544671535491943, Test_Clf_Loss: 2.354465961456299, Test_Filter_Loss: 0.20682205259799957, TEST_Seq_Loss: 0.9799100160598755, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.020357763394713402
Epoch 460, Total Loss: 2.331770181655884, Clf Loss: 2.3317697048187256, Filter Loss: 0.2030186802148819, Seq Loss: 0.9839540719985962, Accuracy Rate: 37.87%, Ones Portion: 0.016292456537485123,             Test_Total_Loss: 2.353966236114502, Test_Clf_Loss: 2.3539650440216064, Test_Filter_Loss: 0.20617079734802246, TEST_Seq_Loss: 0.9799671173095703, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.020308081060647964
Epoch 461, Total Loss: 2.3314766883850098, Clf Loss: 2.3314757347106934, Filter Loss: 0.20161621272563934, Seq Loss: 0.9841076135635376, Accuracy Rate: 37.88%, On

Epoch 479, Total Loss: 2.3304316997528076, Clf Loss: 2.330430507659912, Filter Loss: 0.18321213126182556, Seq Loss: 0.9851037859916687, Accuracy Rate: 37.83%, Ones Portion: 0.015160382725298405,             Test_Total_Loss: 2.35400128364563, Test_Clf_Loss: 2.3539998531341553, Test_Filter_Loss: 0.18673773109912872, TEST_Seq_Loss: 0.981390118598938, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.018931031227111816
Epoch 480, Total Loss: 2.32932710647583, Clf Loss: 2.3293261528015137, Filter Loss: 0.1816847026348114, Seq Loss: 0.9852890968322754, Accuracy Rate: 37.83%, Ones Portion: 0.014985522255301476,             Test_Total_Loss: 2.3547472953796387, Test_Clf_Loss: 2.3547465801239014, Test_Filter_Loss: 0.18512959778308868, TEST_Seq_Loss: 0.9815212488174438, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.01880655251443386
Epoch 481, Total Loss: 2.33172607421875, Clf Loss: 2.3317253589630127, Filter Loss: 0.1801411658525467, Seq Loss: 0.9853441715240479, Accuracy Rate: 37.83%, Ones Po

Epoch 499, Total Loss: 2.332920789718628, Clf Loss: 2.3329200744628906, Filter Loss: 0.15516264736652374, Seq Loss: 0.9866790771484375, Accuracy Rate: 37.83%, Ones Portion: 0.013569677248597145,             Test_Total_Loss: 2.35667085647583, Test_Clf_Loss: 2.3566701412200928, Test_Filter_Loss: 0.1583327203989029, TEST_Seq_Loss: 0.9834381341934204, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.016878101974725723
Epoch 500, Total Loss: 2.3292043209075928, Clf Loss: 2.3292033672332764, Filter Loss: 0.15391063690185547, Seq Loss: 0.9868026971817017, Accuracy Rate: 37.82%, Ones Portion: 0.01344885490834713,             Test_Total_Loss: 2.3563342094421387, Test_Clf_Loss: 2.3563334941864014, Test_Filter_Loss: 0.1569276601076126, TEST_Seq_Loss: 0.9835709929466248, Test_Accuracy_Rate: 37.56%, Test_Ones_Portion: 0.01674865186214447
Epoch 501, Total Loss: 2.3313517570495605, Clf Loss: 2.3313512802124023, Filter Loss: 0.15166287124156952, Seq Loss: 0.9868757724761963, Accuracy Rate: 37.85%, One

Epoch 519, Total Loss: 2.3317489624023438, Clf Loss: 2.3317484855651855, Filter Loss: 0.08563638478517532, Seq Loss: 0.9883260726928711, Accuracy Rate: 37.81%, Ones Portion: 0.011921918019652367,             Test_Total_Loss: 2.355926036834717, Test_Clf_Loss: 2.3559257984161377, Test_Filter_Loss: 0.0859585553407669, TEST_Seq_Loss: 0.9855849146842957, Test_Accuracy_Rate: 37.42%, Test_Ones_Portion: 0.014777732081711292
Epoch 520, Total Loss: 2.3293356895446777, Clf Loss: 2.3293356895446777, Filter Loss: 0.08609174937009811, Seq Loss: 0.9883903861045837, Accuracy Rate: 37.82%, Ones Portion: 0.011861994862556458,             Test_Total_Loss: 2.35676646232605, Test_Clf_Loss: 2.3567662239074707, Test_Filter_Loss: 0.08594626188278198, TEST_Seq_Loss: 0.9855737090110779, Test_Accuracy_Rate: 37.37%, Test_Ones_Portion: 0.014788459055125713
Epoch 521, Total Loss: 2.331045150756836, Clf Loss: 2.331044912338257, Filter Loss: 0.08387483656406403, Seq Loss: 0.9884138703346252, Accuracy Rate: 37.86%, On

Epoch 539, Total Loss: 2.332550048828125, Clf Loss: 2.332549571990967, Filter Loss: 0.05873548984527588, Seq Loss: 0.9887344241142273, Accuracy Rate: 37.83%, Ones Portion: 0.01152421347796917,             Test_Total_Loss: 2.357496738433838, Test_Clf_Loss: 2.3574962615966797, Test_Filter_Loss: 0.05928659439086914, TEST_Seq_Loss: 0.9863520264625549, Test_Accuracy_Rate: 37.37%, Test_Ones_Portion: 0.014010322280228138
Epoch 540, Total Loss: 2.331371545791626, Clf Loss: 2.331371307373047, Filter Loss: 0.05840640887618065, Seq Loss: 0.9886954426765442, Accuracy Rate: 37.85%, Ones Portion: 0.011564207263290882,             Test_Total_Loss: 2.356731414794922, Test_Clf_Loss: 2.3567311763763428, Test_Filter_Loss: 0.059192854911088943, TEST_Seq_Loss: 0.9863536953926086, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.01400800421833992
Epoch 541, Total Loss: 2.3329453468322754, Clf Loss: 2.332944869995117, Filter Loss: 0.056926190853118896, Seq Loss: 0.9888477325439453, Accuracy Rate: 37.86%, One

Epoch 559, Total Loss: 2.332428216934204, Clf Loss: 2.332427740097046, Filter Loss: 0.048781462013721466, Seq Loss: 0.9893207550048828, Accuracy Rate: 37.83%, Ones Portion: 0.01094075571745634,             Test_Total_Loss: 2.3564603328704834, Test_Clf_Loss: 2.3564600944519043, Test_Filter_Loss: 0.050532229244709015, TEST_Seq_Loss: 0.9869996309280396, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.01331110205501318
Epoch 560, Total Loss: 2.331883192062378, Clf Loss: 2.331883192062378, Filter Loss: 0.04859805107116699, Seq Loss: 0.9893736839294434, Accuracy Rate: 37.85%, Ones Portion: 0.01089018676429987,             Test_Total_Loss: 2.356231212615967, Test_Clf_Loss: 2.356231212615967, Test_Filter_Loss: 0.05034210532903671, TEST_Seq_Loss: 0.9870157837867737, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.013292277231812477
Epoch 561, Total Loss: 2.334923267364502, Clf Loss: 2.334923267364502, Filter Loss: 0.04862961918115616, Seq Loss: 0.9893166422843933, Accuracy Rate: 37.82%, Ones 

Epoch 579, Total Loss: 2.33237886428833, Clf Loss: 2.332378625869751, Filter Loss: 0.04583730921149254, Seq Loss: 0.9895668625831604, Accuracy Rate: 37.82%, Ones Portion: 0.01069533172994852,             Test_Total_Loss: 2.35722279548645, Test_Clf_Loss: 2.35722279548645, Test_Filter_Loss: 0.047885626554489136, TEST_Seq_Loss: 0.9872425198554993, Test_Accuracy_Rate: 37.37%, Test_Ones_Portion: 0.013031642884016037
Epoch 580, Total Loss: 2.3314208984375, Clf Loss: 2.331420660018921, Filter Loss: 0.04526323452591896, Seq Loss: 0.9896882176399231, Accuracy Rate: 37.83%, Ones Portion: 0.010562271811068058,             Test_Total_Loss: 2.357450246810913, Test_Clf_Loss: 2.357450008392334, Test_Filter_Loss: 0.04775073006749153, TEST_Seq_Loss: 0.987259566783905, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.013010550290346146
Epoch 581, Total Loss: 2.3323466777801514, Clf Loss: 2.3323464393615723, Filter Loss: 0.04544142633676529, Seq Loss: 0.9896116852760315, Accuracy Rate: 37.82%, Ones Porti

Epoch 599, Total Loss: 2.3321585655212402, Clf Loss: 2.332158327102661, Filter Loss: 0.042322397232055664, Seq Loss: 0.9898597002029419, Accuracy Rate: 37.81%, Ones Portion: 0.010393468663096428,             Test_Total_Loss: 2.3569743633270264, Test_Clf_Loss: 2.356973886489868, Test_Filter_Loss: 0.04501919075846672, TEST_Seq_Loss: 0.9874948859214783, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.012746668420732021
Epoch 600, Total Loss: 2.331798553466797, Clf Loss: 2.3317983150482178, Filter Loss: 0.04201137647032738, Seq Loss: 0.9899615049362183, Accuracy Rate: 37.83%, Ones Portion: 0.010287929326295853,             Test_Total_Loss: 2.3570003509521484, Test_Clf_Loss: 2.3570003509521484, Test_Filter_Loss: 0.0447101853787899, TEST_Seq_Loss: 0.9875140190124512, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.012726063840091228
Epoch 601, Total Loss: 2.3338212966918945, Clf Loss: 2.3338212966918945, Filter Loss: 0.042013466358184814, Seq Loss: 0.9897364974021912, Accuracy Rate: 37.77%

Epoch 619, Total Loss: 2.333317756652832, Clf Loss: 2.333317518234253, Filter Loss: 0.03769650310277939, Seq Loss: 0.990408182144165, Accuracy Rate: 37.80%, Ones Portion: 0.009807443246245384,             Test_Total_Loss: 2.3568787574768066, Test_Clf_Loss: 2.3568787574768066, Test_Filter_Loss: 0.039800014346838, TEST_Seq_Loss: 0.9879908561706543, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.012209479697048664
Epoch 620, Total Loss: 2.3327407836914062, Clf Loss: 2.3327407836914062, Filter Loss: 0.0373750701546669, Seq Loss: 0.9902862310409546, Accuracy Rate: 37.81%, Ones Portion: 0.009925559163093567,             Test_Total_Loss: 2.357700824737549, Test_Clf_Loss: 2.3577001094818115, Test_Filter_Loss: 0.03954039141535759, TEST_Seq_Loss: 0.9880270957946777, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.01217059139162302
Epoch 621, Total Loss: 2.334301233291626, Clf Loss: 2.334300994873047, Filter Loss: 0.037059471011161804, Seq Loss: 0.9902786016464233, Accuracy Rate: 37.76%, Ones 

Epoch 639, Total Loss: 2.3316099643707275, Clf Loss: 2.3316097259521484, Filter Loss: 0.033440034836530685, Seq Loss: 0.9910227656364441, Accuracy Rate: 37.77%, Ones Portion: 0.00915470253676176,             Test_Total_Loss: 2.3581387996673584, Test_Clf_Loss: 2.3581385612487793, Test_Filter_Loss: 0.03578101843595505, TEST_Seq_Loss: 0.9886810183525085, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.011483634822070599
Epoch 640, Total Loss: 2.333122491836548, Clf Loss: 2.333122491836548, Filter Loss: 0.03313639760017395, Seq Loss: 0.991234540939331, Accuracy Rate: 37.76%, Ones Portion: 0.008945402689278126,             Test_Total_Loss: 2.3575940132141113, Test_Clf_Loss: 2.3575937747955322, Test_Filter_Loss: 0.03552176058292389, TEST_Seq_Loss: 0.9887139797210693, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.011451102793216705
Epoch 641, Total Loss: 2.3314154148101807, Clf Loss: 2.3314154148101807, Filter Loss: 0.03309043124318123, Seq Loss: 0.9912581443786621, Accuracy Rate: 37.76%,

Epoch 659, Total Loss: 2.3325705528259277, Clf Loss: 2.3325705528259277, Filter Loss: 0.030819471925497055, Seq Loss: 0.9915924072265625, Accuracy Rate: 37.77%, Ones Portion: 0.00859150942414999,             Test_Total_Loss: 2.3576653003692627, Test_Clf_Loss: 2.3576653003692627, Test_Filter_Loss: 0.032801706343889236, TEST_Seq_Loss: 0.9892110228538513, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.010938994586467743
Epoch 660, Total Loss: 2.332904815673828, Clf Loss: 2.332904577255249, Filter Loss: 0.030397789552807808, Seq Loss: 0.9919188022613525, Accuracy Rate: 37.80%, Ones Portion: 0.008253143168985844,             Test_Total_Loss: 2.3576130867004395, Test_Clf_Loss: 2.3576126098632812, Test_Filter_Loss: 0.032653357833623886, TEST_Seq_Loss: 0.9892380833625793, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.01091096457093954
Epoch 661, Total Loss: 2.3326103687286377, Clf Loss: 2.3326101303100586, Filter Loss: 0.030415168032050133, Seq Loss: 0.9919032454490662, Accuracy Rate: 37.

Epoch 679, Total Loss: 2.336169958114624, Clf Loss: 2.336169719696045, Filter Loss: 0.028229044750332832, Seq Loss: 0.9923741221427917, Accuracy Rate: 37.77%, Ones Portion: 0.007802810054272413,             Test_Total_Loss: 2.357764482498169, Test_Clf_Loss: 2.35776424407959, Test_Filter_Loss: 0.03016059845685959, TEST_Seq_Loss: 0.9896553158760071, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.010489607229828835
Epoch 680, Total Loss: 2.3348655700683594, Clf Loss: 2.3348653316497803, Filter Loss: 0.028116771951317787, Seq Loss: 0.9923617839813232, Accuracy Rate: 37.77%, Ones Portion: 0.007810000330209732,             Test_Total_Loss: 2.3578665256500244, Test_Clf_Loss: 2.3578662872314453, Test_Filter_Loss: 0.0300639308989048, TEST_Seq_Loss: 0.9896649718284607, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.01047773752361536
Epoch 681, Total Loss: 2.335754156112671, Clf Loss: 2.335754156112671, Filter Loss: 0.028026118874549866, Seq Loss: 0.9923834204673767, Accuracy Rate: 37.77%, On

Epoch 699, Total Loss: 2.3346900939941406, Clf Loss: 2.3346898555755615, Filter Loss: 0.02644868567585945, Seq Loss: 0.9926889538764954, Accuracy Rate: 37.75%, Ones Portion: 0.007479680236428976,             Test_Total_Loss: 2.358268976211548, Test_Clf_Loss: 2.3582687377929688, Test_Filter_Loss: 0.028267107903957367, TEST_Seq_Loss: 0.9900122284889221, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.010141746141016483
Epoch 700, Total Loss: 2.335869073867798, Clf Loss: 2.3358688354492188, Filter Loss: 0.026359137147665024, Seq Loss: 0.9926868081092834, Accuracy Rate: 37.75%, Ones Portion: 0.007479036692529917,             Test_Total_Loss: 2.358086585998535, Test_Clf_Loss: 2.358086109161377, Test_Filter_Loss: 0.02817562408745289, TEST_Seq_Loss: 0.9900445342063904, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.010107416659593582
Epoch 701, Total Loss: 2.333141326904297, Clf Loss: 2.3331408500671387, Filter Loss: 0.02612530067563057, Seq Loss: 0.9928569793701172, Accuracy Rate: 37.74%,

Epoch 719, Total Loss: 2.333808422088623, Clf Loss: 2.333808422088623, Filter Loss: 0.02475251629948616, Seq Loss: 0.9929782748222351, Accuracy Rate: 37.76%, Ones Portion: 0.00718527240678668,             Test_Total_Loss: 2.3569393157958984, Test_Clf_Loss: 2.3569390773773193, Test_Filter_Loss: 0.026262085884809494, TEST_Seq_Loss: 0.9904562830924988, Test_Accuracy_Rate: 37.42%, Test_Ones_Portion: 0.009705580770969391
Epoch 720, Total Loss: 2.334839344024658, Clf Loss: 2.3348388671875, Filter Loss: 0.024342728778719902, Seq Loss: 0.9932242631912231, Accuracy Rate: 37.75%, Ones Portion: 0.006932706572115421,             Test_Total_Loss: 2.356987476348877, Test_Clf_Loss: 2.356987237930298, Test_Filter_Loss: 0.026135168969631195, TEST_Seq_Loss: 0.9904771447181702, Test_Accuracy_Rate: 37.42%, Test_Ones_Portion: 0.009685725905001163
Epoch 721, Total Loss: 2.333827495574951, Clf Loss: 2.333827018737793, Filter Loss: 0.024204745888710022, Seq Loss: 0.993202805519104, Accuracy Rate: 37.76%, Ones

Epoch 739, Total Loss: 2.3362345695495605, Clf Loss: 2.3362345695495605, Filter Loss: 0.02239834889769554, Seq Loss: 0.9934973120689392, Accuracy Rate: 37.76%, Ones Portion: 0.00664186105132103,             Test_Total_Loss: 2.3581883907318115, Test_Clf_Loss: 2.3581881523132324, Test_Filter_Loss: 0.024090444669127464, TEST_Seq_Loss: 0.9909510016441345, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.009211898781359196
Epoch 740, Total Loss: 2.3321094512939453, Clf Loss: 2.3321094512939453, Filter Loss: 0.022346334531903267, Seq Loss: 0.9935593605041504, Accuracy Rate: 37.75%, Ones Portion: 0.006574900355190039,             Test_Total_Loss: 2.3584206104278564, Test_Clf_Loss: 2.3584203720092773, Test_Filter_Loss: 0.024002768099308014, TEST_Seq_Loss: 0.9909675717353821, Test_Accuracy_Rate: 37.32%, Test_Ones_Portion: 0.00919247791171074
Epoch 741, Total Loss: 2.3318252563476562, Clf Loss: 2.3318252563476562, Filter Loss: 0.022221770137548447, Seq Loss: 0.993524432182312, Accuracy Rate: 37.

Epoch 759, Total Loss: 2.3348546028137207, Clf Loss: 2.3348546028137207, Filter Loss: 0.02126932330429554, Seq Loss: 0.9936680793762207, Accuracy Rate: 37.80%, Ones Portion: 0.006460628006607294,             Test_Total_Loss: 2.3582918643951416, Test_Clf_Loss: 2.3582916259765625, Test_Filter_Loss: 0.022913532331585884, TEST_Seq_Loss: 0.9911866784095764, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.008966165594756603
Epoch 760, Total Loss: 2.3338849544525146, Clf Loss: 2.3338849544525146, Filter Loss: 0.02109537273645401, Seq Loss: 0.993860125541687, Accuracy Rate: 37.81%, Ones Portion: 0.006262833718210459,             Test_Total_Loss: 2.3583624362945557, Test_Clf_Loss: 2.3583621978759766, Test_Filter_Loss: 0.02283652499318123, TEST_Seq_Loss: 0.9912090301513672, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.008942109532654285
Epoch 761, Total Loss: 2.333326816558838, Clf Loss: 2.333326816558838, Filter Loss: 0.02105766162276268, Seq Loss: 0.9937264323234558, Accuracy Rate: 37.81%

Epoch 779, Total Loss: 2.333793878555298, Clf Loss: 2.333793878555298, Filter Loss: 0.0195919256657362, Seq Loss: 0.9941148161888123, Accuracy Rate: 37.81%, Ones Portion: 0.005997805390506983,             Test_Total_Loss: 2.358363151550293, Test_Clf_Loss: 2.358363151550293, Test_Filter_Loss: 0.021404441446065903, TEST_Seq_Loss: 0.9915387630462646, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.008613316342234612
Epoch 780, Total Loss: 2.333679676055908, Clf Loss: 2.333679676055908, Filter Loss: 0.01966136135160923, Seq Loss: 0.9938602447509766, Accuracy Rate: 37.81%, Ones Portion: 0.006258613895624876,             Test_Total_Loss: 2.3582375049591064, Test_Clf_Loss: 2.3582372665405273, Test_Filter_Loss: 0.02131469175219536, TEST_Seq_Loss: 0.9915509819984436, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.008600970730185509
Epoch 781, Total Loss: 2.3336691856384277, Clf Loss: 2.3336691856384277, Filter Loss: 0.019441032782197, Seq Loss: 0.9940494298934937, Accuracy Rate: 37.81%, Ones

Epoch 799, Total Loss: 2.3311586380004883, Clf Loss: 2.3311586380004883, Filter Loss: 0.018195893615484238, Seq Loss: 0.9943822026252747, Accuracy Rate: 37.81%, Ones Portion: 0.005723627749830484,             Test_Total_Loss: 2.3583364486694336, Test_Clf_Loss: 2.3583364486694336, Test_Filter_Loss: 0.020069483667612076, TEST_Seq_Loss: 0.991854190826416, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.008301297202706337
Epoch 800, Total Loss: 2.3345603942871094, Clf Loss: 2.3345603942871094, Filter Loss: 0.01828940398991108, Seq Loss: 0.9942256212234497, Accuracy Rate: 37.81%, Ones Portion: 0.005878108087927103,             Test_Total_Loss: 2.3585329055786133, Test_Clf_Loss: 2.3585329055786133, Test_Filter_Loss: 0.02002439647912979, TEST_Seq_Loss: 0.9918637275695801, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.00829220749437809
Epoch 801, Total Loss: 2.334625005722046, Clf Loss: 2.334625005722046, Filter Loss: 0.01829267479479313, Seq Loss: 0.9943791031837463, Accuracy Rate: 37.81%

Epoch 819, Total Loss: 2.3344650268554688, Clf Loss: 2.3344650268554688, Filter Loss: 0.017060192301869392, Seq Loss: 0.994513750076294, Accuracy Rate: 37.82%, Ones Portion: 0.005586941260844469,             Test_Total_Loss: 2.358363628387451, Test_Clf_Loss: 2.358363628387451, Test_Filter_Loss: 0.018723608925938606, TEST_Seq_Loss: 0.9922152161598206, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.007952277548611164
Epoch 820, Total Loss: 2.332836151123047, Clf Loss: 2.332836151123047, Filter Loss: 0.016885899007320404, Seq Loss: 0.9944563508033752, Accuracy Rate: 37.82%, Ones Portion: 0.005643583368510008,             Test_Total_Loss: 2.358707904815674, Test_Clf_Loss: 2.358707904815674, Test_Filter_Loss: 0.018637923523783684, TEST_Seq_Loss: 0.9922332763671875, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.007934533059597015
Epoch 821, Total Loss: 2.3343138694763184, Clf Loss: 2.3343138694763184, Filter Loss: 0.016749771311879158, Seq Loss: 0.9945372343063354, Accuracy Rate: 37.82%

Epoch 839, Total Loss: 2.33429217338562, Clf Loss: 2.33429217338562, Filter Loss: 0.015532433055341244, Seq Loss: 0.9946706891059875, Accuracy Rate: 37.82%, Ones Portion: 0.005428781732916832,             Test_Total_Loss: 2.3584184646606445, Test_Clf_Loss: 2.3584184646606445, Test_Filter_Loss: 0.017222071066498756, TEST_Seq_Loss: 0.9926138520240784, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.00755819957703352
Epoch 840, Total Loss: 2.3326334953308105, Clf Loss: 2.3326334953308105, Filter Loss: 0.015344465151429176, Seq Loss: 0.9947186708450317, Accuracy Rate: 37.82%, Ones Portion: 0.005378182977437973,             Test_Total_Loss: 2.358447790145874, Test_Clf_Loss: 2.358447790145874, Test_Filter_Loss: 0.017171399667859077, TEST_Seq_Loss: 0.9926252961158752, Test_Accuracy_Rate: 37.27%, Test_Ones_Portion: 0.007545815780758858
Epoch 841, Total Loss: 2.3346879482269287, Clf Loss: 2.3346879482269287, Filter Loss: 0.015355619601905346, Seq Loss: 0.994732677936554, Accuracy Rate: 37.82%,

Epoch 859, Total Loss: 2.3340561389923096, Clf Loss: 2.3340561389923096, Filter Loss: 0.014228650368750095, Seq Loss: 0.9949349164962769, Accuracy Rate: 37.80%, Ones Portion: 0.005175907630473375,             Test_Total_Loss: 2.3588171005249023, Test_Clf_Loss: 2.3588171005249023, Test_Filter_Loss: 0.015993880107998848, TEST_Seq_Loss: 0.9929651618003845, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.00718859676271677
Epoch 860, Total Loss: 2.334104299545288, Clf Loss: 2.334104299545288, Filter Loss: 0.014193614944815636, Seq Loss: 0.9950583577156067, Accuracy Rate: 37.80%, Ones Portion: 0.005052295513451099,             Test_Total_Loss: 2.3592803478240967, Test_Clf_Loss: 2.3592803478240967, Test_Filter_Loss: 0.015926508232951164, TEST_Seq_Loss: 0.9929919242858887, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.007160373497754335
Epoch 861, Total Loss: 2.3331968784332275, Clf Loss: 2.3331968784332275, Filter Loss: 0.01407941896468401, Seq Loss: 0.995079755783081, Accuracy Rate: 37.8

Epoch 879, Total Loss: 2.3330485820770264, Clf Loss: 2.3330485820770264, Filter Loss: 0.01324137020856142, Seq Loss: 0.9952046275138855, Accuracy Rate: 37.81%, Ones Portion: 0.00490847323089838,             Test_Total_Loss: 2.358353614807129, Test_Clf_Loss: 2.358353614807129, Test_Filter_Loss: 0.014948999509215355, TEST_Seq_Loss: 0.9932411313056946, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006876092869788408
Epoch 880, Total Loss: 2.3351268768310547, Clf Loss: 2.3351268768310547, Filter Loss: 0.013276699930429459, Seq Loss: 0.9951953887939453, Accuracy Rate: 37.80%, Ones Portion: 0.004916703328490257,             Test_Total_Loss: 2.3586394786834717, Test_Clf_Loss: 2.3586394786834717, Test_Filter_Loss: 0.014903531409800053, TEST_Seq_Loss: 0.9932544827461243, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006863366346806288
Epoch 881, Total Loss: 2.337585926055908, Clf Loss: 2.337585926055908, Filter Loss: 0.013109187595546246, Seq Loss: 0.9952850341796875, Accuracy Rate: 37.78

Epoch 899, Total Loss: 2.3344523906707764, Clf Loss: 2.3344523906707764, Filter Loss: 0.012445646338164806, Seq Loss: 0.9954326748847961, Accuracy Rate: 37.80%, Ones Portion: 0.004669905640184879,             Test_Total_Loss: 2.3585355281829834, Test_Clf_Loss: 2.3585355281829834, Test_Filter_Loss: 0.01420845277607441, TEST_Seq_Loss: 0.993408203125, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.0066870409063994884
Epoch 900, Total Loss: 2.3357999324798584, Clf Loss: 2.3357999324798584, Filter Loss: 0.012443208135664463, Seq Loss: 0.9954229593276978, Accuracy Rate: 37.80%, Ones Portion: 0.004681658931076527,             Test_Total_Loss: 2.3591063022613525, Test_Clf_Loss: 2.3591063022613525, Test_Filter_Loss: 0.014174863696098328, TEST_Seq_Loss: 0.9934160709381104, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.0066788154654204845
Epoch 901, Total Loss: 2.332106590270996, Clf Loss: 2.332106590270996, Filter Loss: 0.012364743277430534, Seq Loss: 0.9954891204833984, Accuracy Rate: 37.8

Epoch 919, Total Loss: 2.336170196533203, Clf Loss: 2.336170196533203, Filter Loss: 0.011861252598464489, Seq Loss: 0.9956466555595398, Accuracy Rate: 37.80%, Ones Portion: 0.004451273009181023,             Test_Total_Loss: 2.358567714691162, Test_Clf_Loss: 2.358567714691162, Test_Filter_Loss: 0.013666362501680851, TEST_Seq_Loss: 0.9935399293899536, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006546339951455593
Epoch 920, Total Loss: 2.33335280418396, Clf Loss: 2.33335280418396, Filter Loss: 0.011823805049061775, Seq Loss: 0.9955930113792419, Accuracy Rate: 37.80%, Ones Portion: 0.004505795892328024,             Test_Total_Loss: 2.3595540523529053, Test_Clf_Loss: 2.3595540523529053, Test_Filter_Loss: 0.0136445052921772, TEST_Seq_Loss: 0.9935455322265625, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006541616283357143
Epoch 921, Total Loss: 2.3358030319213867, Clf Loss: 2.3358030319213867, Filter Loss: 0.01191126648336649, Seq Loss: 0.9955177903175354, Accuracy Rate: 37.80%, On

Epoch 939, Total Loss: 2.334946632385254, Clf Loss: 2.334946632385254, Filter Loss: 0.011505265720188618, Seq Loss: 0.9956840872764587, Accuracy Rate: 37.80%, Ones Portion: 0.004412730224430561,             Test_Total_Loss: 2.3585445880889893, Test_Clf_Loss: 2.3585445880889893, Test_Filter_Loss: 0.01329025812447071, TEST_Seq_Loss: 0.9936238527297974, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006458301562815905
Epoch 940, Total Loss: 2.332343816757202, Clf Loss: 2.332343816757202, Filter Loss: 0.011556850746273994, Seq Loss: 0.9955581426620483, Accuracy Rate: 37.80%, Ones Portion: 0.004543574992567301,             Test_Total_Loss: 2.3590874671936035, Test_Clf_Loss: 2.3590874671936035, Test_Filter_Loss: 0.013268765062093735, TEST_Seq_Loss: 0.9936299324035645, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006452689878642559
Epoch 941, Total Loss: 2.3335752487182617, Clf Loss: 2.3335752487182617, Filter Loss: 0.011482498608529568, Seq Loss: 0.9957265853881836, Accuracy Rate: 37.8

Epoch 959, Total Loss: 2.3338518142700195, Clf Loss: 2.3338518142700195, Filter Loss: 0.011260662227869034, Seq Loss: 0.9957576990127563, Accuracy Rate: 37.78%, Ones Portion: 0.004335989244282246,             Test_Total_Loss: 2.3593931198120117, Test_Clf_Loss: 2.3593931198120117, Test_Filter_Loss: 0.012970478273928165, TEST_Seq_Loss: 0.9937078952789307, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006372148171067238
Epoch 960, Total Loss: 2.3348846435546875, Clf Loss: 2.3348846435546875, Filter Loss: 0.01118200272321701, Seq Loss: 0.9958040714263916, Accuracy Rate: 37.78%, Ones Portion: 0.0042881653644144535,             Test_Total_Loss: 2.358466625213623, Test_Clf_Loss: 2.358466625213623, Test_Filter_Loss: 0.012958372943103313, TEST_Seq_Loss: 0.9937093257904053, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006370609160512686
Epoch 961, Total Loss: 2.334662437438965, Clf Loss: 2.334662437438965, Filter Loss: 0.01120404340326786, Seq Loss: 0.9957334995269775, Accuracy Rate: 37.7

Epoch 979, Total Loss: 2.336907386779785, Clf Loss: 2.336907386779785, Filter Loss: 0.011049163527786732, Seq Loss: 0.9958274364471436, Accuracy Rate: 37.78%, Ones Portion: 0.00427172752097249,             Test_Total_Loss: 2.358797550201416, Test_Clf_Loss: 2.358797550201416, Test_Filter_Loss: 0.01273647416383028, TEST_Seq_Loss: 0.9937765598297119, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006304103881120682
Epoch 980, Total Loss: 2.332771062850952, Clf Loss: 2.332771062850952, Filter Loss: 0.010973702184855938, Seq Loss: 0.9957985281944275, Accuracy Rate: 37.78%, Ones Portion: 0.004302294924855232,             Test_Total_Loss: 2.3584556579589844, Test_Clf_Loss: 2.3584556579589844, Test_Filter_Loss: 0.012726468965411186, TEST_Seq_Loss: 0.9937774538993835, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.0063033695332705975
Epoch 981, Total Loss: 2.333857297897339, Clf Loss: 2.333857297897339, Filter Loss: 0.011008167639374733, Seq Loss: 0.9957888126373291, Accuracy Rate: 37.78%, 

Epoch 999, Total Loss: 2.3361940383911133, Clf Loss: 2.3361940383911133, Filter Loss: 0.010756785050034523, Seq Loss: 0.9959046840667725, Accuracy Rate: 37.78%, Ones Portion: 0.004193978384137154,             Test_Total_Loss: 2.3596694469451904, Test_Clf_Loss: 2.3596694469451904, Test_Filter_Loss: 0.012512125074863434, TEST_Seq_Loss: 0.9938420653343201, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006242213770747185
Epoch 1000, Total Loss: 2.3341102600097656, Clf Loss: 2.3341102600097656, Filter Loss: 0.010674076154828072, Seq Loss: 0.9960187673568726, Accuracy Rate: 37.78%, Ones Portion: 0.004078774247318506,             Test_Total_Loss: 2.359602451324463, Test_Clf_Loss: 2.359602451324463, Test_Filter_Loss: 0.012498367577791214, TEST_Seq_Loss: 0.9938464164733887, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006237739231437445
Epoch 1001, Total Loss: 2.3307313919067383, Clf Loss: 2.3307313919067383, Filter Loss: 0.010814176872372627, Seq Loss: 0.9958400726318359, Accuracy Rate:

Epoch 1019, Total Loss: 2.3347089290618896, Clf Loss: 2.3347089290618896, Filter Loss: 0.01055168453603983, Seq Loss: 0.9960058331489563, Accuracy Rate: 37.78%, Ones Portion: 0.004092786926776171,             Test_Total_Loss: 2.3586983680725098, Test_Clf_Loss: 2.3586983680725098, Test_Filter_Loss: 0.012292778119444847, TEST_Seq_Loss: 0.9938989281654358, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.00618875864893198
Epoch 1020, Total Loss: 2.336052417755127, Clf Loss: 2.336052417755127, Filter Loss: 0.01056584157049656, Seq Loss: 0.9959753155708313, Accuracy Rate: 37.78%, Ones Portion: 0.004114643670618534,             Test_Total_Loss: 2.3598480224609375, Test_Clf_Loss: 2.3598480224609375, Test_Filter_Loss: 0.012282459996640682, TEST_Seq_Loss: 0.9939011931419373, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006186773534864187
Epoch 1021, Total Loss: 2.3337602615356445, Clf Loss: 2.3337602615356445, Filter Loss: 0.010515621863305569, Seq Loss: 0.9960496425628662, Accuracy Rate: 3

Epoch 1039, Total Loss: 2.3337554931640625, Clf Loss: 2.3337554931640625, Filter Loss: 0.010349583812057972, Seq Loss: 0.9960127472877502, Accuracy Rate: 37.78%, Ones Portion: 0.004092635586857796,             Test_Total_Loss: 2.358684778213501, Test_Clf_Loss: 2.358684778213501, Test_Filter_Loss: 0.012093986384570599, TEST_Seq_Loss: 0.9939642548561096, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006128851789981127
Epoch 1040, Total Loss: 2.3331899642944336, Clf Loss: 2.3331899642944336, Filter Loss: 0.010310023091733456, Seq Loss: 0.9961065649986267, Accuracy Rate: 37.78%, Ones Portion: 0.003996004816144705,             Test_Total_Loss: 2.3586928844451904, Test_Clf_Loss: 2.3586928844451904, Test_Filter_Loss: 0.012090369127690792, TEST_Seq_Loss: 0.9939670562744141, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006126906257122755
Epoch 1041, Total Loss: 2.3351352214813232, Clf Loss: 2.3351352214813232, Filter Loss: 0.010357312858104706, Seq Loss: 0.9960448145866394, Accuracy Rate

Epoch 1059, Total Loss: 2.3356428146362305, Clf Loss: 2.3356428146362305, Filter Loss: 0.010164589621126652, Seq Loss: 0.9962009191513062, Accuracy Rate: 37.78%, Ones Portion: 0.003902066731825471,             Test_Total_Loss: 2.358595609664917, Test_Clf_Loss: 2.358595609664917, Test_Filter_Loss: 0.01190932560712099, TEST_Seq_Loss: 0.9940322637557983, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006069279741495848
Epoch 1060, Total Loss: 2.3350281715393066, Clf Loss: 2.3350281715393066, Filter Loss: 0.010188672691583633, Seq Loss: 0.9960384368896484, Accuracy Rate: 37.78%, Ones Portion: 0.004061130341142416,             Test_Total_Loss: 2.358694076538086, Test_Clf_Loss: 2.358694076538086, Test_Filter_Loss: 0.011904114857316017, TEST_Seq_Loss: 0.9940352439880371, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006066585890948772
Epoch 1061, Total Loss: 2.335278272628784, Clf Loss: 2.335278272628784, Filter Loss: 0.010198775678873062, Seq Loss: 0.9962023496627808, Accuracy Rate: 37.

Epoch 1079, Total Loss: 2.3345773220062256, Clf Loss: 2.3345773220062256, Filter Loss: 0.010041311383247375, Seq Loss: 0.9962198138237, Accuracy Rate: 37.78%, Ones Portion: 0.0038939921651035547,             Test_Total_Loss: 2.3586227893829346, Test_Clf_Loss: 2.3586227893829346, Test_Filter_Loss: 0.011789253912866116, TEST_Seq_Loss: 0.9940934777259827, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006013978738337755
Epoch 1080, Total Loss: 2.335097551345825, Clf Loss: 2.335097551345825, Filter Loss: 0.010174225084483624, Seq Loss: 0.9960899949073792, Accuracy Rate: 37.78%, Ones Portion: 0.004023573361337185,             Test_Total_Loss: 2.358281135559082, Test_Clf_Loss: 2.358281135559082, Test_Filter_Loss: 0.011784911155700684, TEST_Seq_Loss: 0.9940953254699707, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.006012406665831804
Epoch 1081, Total Loss: 2.332092046737671, Clf Loss: 2.332092046737671, Filter Loss: 0.01007442083209753, Seq Loss: 0.9961181879043579, Accuracy Rate: 37.77

Epoch 1099, Total Loss: 2.3355183601379395, Clf Loss: 2.3355183601379395, Filter Loss: 0.009964934550225735, Seq Loss: 0.9962552785873413, Accuracy Rate: 37.77%, Ones Portion: 0.00385313224978745,             Test_Total_Loss: 2.3581106662750244, Test_Clf_Loss: 2.3581106662750244, Test_Filter_Loss: 0.011687381193041801, TEST_Seq_Loss: 0.9941492676734924, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005965784657746553
Epoch 1100, Total Loss: 2.334296703338623, Clf Loss: 2.334296703338623, Filter Loss: 0.009936338290572166, Seq Loss: 0.9961929321289062, Accuracy Rate: 37.77%, Ones Portion: 0.003920367453247309,             Test_Total_Loss: 2.359208583831787, Test_Clf_Loss: 2.359208583831787, Test_Filter_Loss: 0.011680074036121368, TEST_Seq_Loss: 0.9941515922546387, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005963724106550217
Epoch 1101, Total Loss: 2.337137222290039, Clf Loss: 2.337137222290039, Filter Loss: 0.009978820569813251, Seq Loss: 0.996185302734375, Accuracy Rate: 37.7

Epoch 1119, Total Loss: 2.3337242603302, Clf Loss: 2.3337242603302, Filter Loss: 0.009768160991370678, Seq Loss: 0.9963800311088562, Accuracy Rate: 37.77%, Ones Portion: 0.0037338952533900738,             Test_Total_Loss: 2.358189344406128, Test_Clf_Loss: 2.358189344406128, Test_Filter_Loss: 0.01157358754426241, TEST_Seq_Loss: 0.9942164421081543, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005906575825065374
Epoch 1120, Total Loss: 2.335392713546753, Clf Loss: 2.335392713546753, Filter Loss: 0.009800074622035027, Seq Loss: 0.996300458908081, Accuracy Rate: 37.77%, Ones Portion: 0.0038074390031397343,             Test_Total_Loss: 2.3600118160247803, Test_Clf_Loss: 2.3600118160247803, Test_Filter_Loss: 0.011569072492420673, TEST_Seq_Loss: 0.9942231774330139, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005901684984564781
Epoch 1121, Total Loss: 2.334084987640381, Clf Loss: 2.334084987640381, Filter Loss: 0.009856391698122025, Seq Loss: 0.9962635636329651, Accuracy Rate: 37.77%, 

Epoch 1139, Total Loss: 2.335972785949707, Clf Loss: 2.335972785949707, Filter Loss: 0.00967133417725563, Seq Loss: 0.996351957321167, Accuracy Rate: 37.77%, Ones Portion: 0.0037574267480522394,             Test_Total_Loss: 2.359710931777954, Test_Clf_Loss: 2.359710931777954, Test_Filter_Loss: 0.011456304229795933, TEST_Seq_Loss: 0.9942799806594849, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005850082729011774
Epoch 1140, Total Loss: 2.3366289138793945, Clf Loss: 2.3366289138793945, Filter Loss: 0.009856757707893848, Seq Loss: 0.9961929321289062, Accuracy Rate: 37.77%, Ones Portion: 0.003921377472579479,             Test_Total_Loss: 2.3595032691955566, Test_Clf_Loss: 2.3595032691955566, Test_Filter_Loss: 0.011445284821093082, TEST_Seq_Loss: 0.9942823052406311, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005848264787346125
Epoch 1141, Total Loss: 2.3328773975372314, Clf Loss: 2.3328773975372314, Filter Loss: 0.009683218784630299, Seq Loss: 0.996364176273346, Accuracy Rate: 37

Epoch 1159, Total Loss: 2.336899518966675, Clf Loss: 2.336899518966675, Filter Loss: 0.009624398313462734, Seq Loss: 0.9963691234588623, Accuracy Rate: 37.77%, Ones Portion: 0.003746556118130684,             Test_Total_Loss: 2.3599796295166016, Test_Clf_Loss: 2.3599796295166016, Test_Filter_Loss: 0.011342810466885567, TEST_Seq_Loss: 0.9943491816520691, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.005786838475614786
Epoch 1160, Total Loss: 2.3370211124420166, Clf Loss: 2.3370211124420166, Filter Loss: 0.00954380165785551, Seq Loss: 0.9963944554328918, Accuracy Rate: 37.77%, Ones Portion: 0.003723644884303212,             Test_Total_Loss: 2.3596880435943604, Test_Clf_Loss: 2.3596880435943604, Test_Filter_Loss: 0.011335517279803753, TEST_Seq_Loss: 0.9943549036979675, Test_Accuracy_Rate: 37.22%, Test_Ones_Portion: 0.0057812705636024475
Epoch 1161, Total Loss: 2.337895154953003, Clf Loss: 2.337895154953003, Filter Loss: 0.009555195458233356, Seq Loss: 0.9964841604232788, Accuracy Rate: 

### Store process

In [ ]:
save_dir = './results/'+DateID+'/'
if not os.path.isdir(save_dir):
    os.makedirs(save_dir, exist_ok=True)
saveR = save_dir + 'losses_metrics.xlsx'
data = {'train total loss':train_weighted_loss, 'train acc loss':train_loss_acc,
        'train filter loss':train_loss_filter,'train seq loss':train_loss_seq,
        'train acc rate':train_acc_rate, 'train ones num':train_ones_num,
        'test total loss':test_weighted_loss, 'test acc loss':test_loss_acc,
        'test filter loss':test_loss_filter, 'test seq loss': test_loss_seq,
        'test acc rate':test_acc_rate, 'test ones num':test_ones_num
       }
df = pd.DataFrame(data)
df.to_excel(saveR)
print(saveR)
df

## Inference
* Case study

In [ ]:
labelID_dict = {'copper': 6, 'livestock': 28, 'gold': 25, 'money-fx': 19, 'ipi': 30, 'trade': 11, 
                'cocoa': 0, 'iron-steel': 31, 'reserves': 12, 'tin': 26, 'zinc': 37, 'jobs': 34, 
                'ship': 13, 'cotton': 14, 'alum': 23, 'strategic-metal': 27, 'lead': 45, 
                'housing': 7, 'meal-feed': 22, 'gnp': 21, 'sugar': 10, 'rubber': 32, 'dlr': 40,
                'veg-oil': 2, 'interest': 20, 'crude': 16, 'coffee': 9, 'wheat': 5, 'carcass': 15,
                'lei': 35, 'gas': 41, 'nat-gas': 17, 'oilseed': 24, 'orange': 38,
                'heat': 33, 'wpi': 43, 'silver': 42, 'cpi': 18, 'earn': 3, 'bop': 36,
                'money-supply': 8, 'hog': 44, 'acq': 4, 'pet-chem': 39, 'grain': 1, 'retail': 29} # 46個cat
IDlabel_dict = {v: k for k, v in labelID_dict.items()}
word_idx_pd = pd.DataFrame(data=word_index,index=["ID"]).T
word_idx_pd['ori_word'] = word_idx_pd.index 
word_idx_pd = word_idx_pd.set_index(["ID"])
print(word_idx_pd)

def getKeysByValues(listOfValues,dfOfElements=word_idx_pd):
    ori_word_li = []
    for item in listOfValues:
        key = item
#         print(key)
        if key == 0:
            continue
        ori_word = dfOfElements.loc[key][0]
        ori_word_li.append(ori_word)
    return ori_word_li

In [ ]:
word_fil_text = []
word_ori_text = []
from tqdm import tqdm
for text in tqdm(int_ori_text): #text = one epoch
    temp = []
    for t in text: #t = one sample
        temp.append(getKeysByValues(list(t)))
    word_ori_text.append(temp)
for text in tqdm(int_fil_text):
    temp = []
    for t in text:
        temp.append(getKeysByValues(list(t)))
    word_fil_text.append(temp)

save_path = save_dir + "original_words.pkl"
pickle.dump(file=open(save_path,'wb'),obj=word_ori_text)
save_path = save_dir + "filterUsed_words.pkl"
pickle.dump(file=open(save_path,'wb'),obj=word_fil_text)
print(save_path)

In [ ]:
col = ['epoch#','text_id','ori_text','filter_text','label']
df = pd.DataFrame(columns=col)
for epoch,(text_ori,text_fil,labelID) in tqdm(enumerate(zip(word_ori_text,word_fil_text,int_label))):
    for idx,(t_o,t_f,label) in enumerate(zip(text_ori,text_fil,labelID)):
        temp = pd.Series([epoch,idx," ".join(t_o)," ".join(t_f),IDlabel_dict[label]],col)
        df = df.append(temp,ignore_index=True)

save_path = save_dir + "caseStudy_result.xlsx"
df.to_excel(save_path,index=False)
print(save_path)
df

4

# Deprecated

In [161]:
save_path = save_dir + "caseStudy_result.xlsx"
df.to_excel(save_path,index=False)
save_path = save_dir + "original_words.pkl"
pickle.dump(file=open(save_path,'wb'),obj=word_ori_text)
save_path = save_dir + "filterUsed_words.pkl"
pickle.dump(file=open(save_path,'wb'),obj=word_fil_text)

In [ ]:
def test_step(x,yc):
    emb, pred_imp = model1(x)
    #loss1 = alpha*loss_object1(pred_imp) #phase1
    #pred_imp2 = tf.math.round(pred_imp)
    #pred_imp3 = tf.clip_by_value(pred_imp,clip_value_max=1,clip_value_min=0)
    pred_imp2 = tf.math.round(pred_imp)
    loss1 = loss_object1(pred_imp) #phase2
    pred_cat = model2(emb,pred_imp2) #phase1: pred_imp ; phase2:pred_imp2
    loss2 = loss_object2(yc, pred_cat)
    loss3 = 1-seq_loss(pred_imp2)
    #t_loss = loss1 + loss2
    t_loss = alpha*loss1 + beta*loss2 + gamma*loss3
    
    test_loss(t_loss)
    test_filloss(loss1)
    test_accloss(loss2)
    test_seqloss(loss3)
    test_accuracy(yc, pred_cat)
    t_ones = one_percentage(pred_imp2) #pred_imp2
    test_ones(t_ones)
for test_text, test_labels in valid_ds:
    test_step(test_text, test_labels)

In [7]:
for test_text, test_labels in valid_ds:
    print(test_text)
    print(test_labels)
    break
#     test_step(test_text, test_labels)

NameError: name 'valid_ds' is not defined

In [11]:
convert_li = getKeysByValues(testX[0])
convert_li

['the',
 'in',
 'compared',
 'analysts',
 'etl',
 'at',
 'may',
 'said',
 'si',
 'offshore',
 'a',
 'but',
 'a',
 'ohio',
 'and',
 'publicly',
 'its',
 'sent',
 '380',
 'plan',
 'bank',
 'mln',
 '1',
 '756',
 'not',
 'note',
 'was',
 'finding',
 'eased',
 '80',
 'in',
 'compared',
 'analysts',
 'as',
 'main',
 'and',
 'origins',
 'be',
 'in',
 'borrowers',
 'said',
 'in',
 'hearings',
 'said',
 'becoming',
 '3',
 'as',
 'of',
 'was',
 'in',
 'worsen',
 'analysts',
 'that',
 'of',
 'offshore',
 'flight',
 'for',
 'during',
 'time',
 'official',
 'mln',
 'a',
 'for',
 'capital',
 'in',
 'par',
 'said',
 'in',
 'worsen',
 'sought',
 'bank',
 'sales',
 'commodity',
 'common',
 'above',
 'a',
 'merchants',
 'amounted',
 'already',
 'in',
 'capital',
 'mln',
 'in',
 'compared',
 'analysts',
 '3',
 'in',
 'worsen',
 '87',
 '41',
 'for',
 'hearings',
 'said',
 'becoming',
 'importers',
 'offshore',
 'flight',
 'be',
 'in',
 'has',
 'would',
 '24',
 '32',
 'firms',
 'of',
 'a',
 'only',
 'cattl

In [12]:
reuters.get_word_index()

{'mdbl': 10996,
 'fawc': 16260,
 'degussa': 12089,
 'woods': 8803,
 'hanging': 13796,
 'localized': 20672,
 'sation': 20673,
 'chanthaburi': 20675,
 'refunding': 10997,
 'hermann': 8804,
 'passsengers': 20676,
 'stipulate': 20677,
 'heublein': 8352,
 'screaming': 20713,
 'tcby': 16261,
 'four': 185,
 'grains': 1642,
 'broiler': 20680,
 'wooden': 12090,
 'wednesday': 1220,
 'highveld': 13797,
 'duffour': 7593,
 '0053': 20681,
 'elections': 3914,
 '270': 2563,
 '271': 3551,
 '272': 5113,
 '273': 3552,
 '274': 3400,
 'rudman': 7975,
 '276': 3401,
 '277': 3478,
 '278': 3632,
 '279': 4309,
 'dormancy': 9381,
 'errors': 7247,
 'deferred': 3086,
 'sptnd': 20683,
 'cooking': 8805,
 'stratabit': 20684,
 'designing': 16262,
 'metalurgicos': 20685,
 'databank': 13798,
 '300er': 20686,
 'shocks': 20687,
 'nawg': 7972,
 'tnta': 20688,
 'perforations': 20689,
 'affiliates': 2891,
 '27p': 20690,
 'ching': 16263,
 'china': 595,
 'wagyu': 16264,
 'affiliated': 3189,
 'chino': 16265,
 'chinh': 16266,
 '

ori_word
ID              
10996       mdbl
16260       fawc
12089    degussa
8803       woods
13796    hanging
...          ...
16258    rotting
10995       pods
2849       emery
30979  northerly
16259   onomichi

[30979 rows x 1 columns]

In [1047]:
test_text[0][0].numpy()

1

In [1049]:
word_idx_pd.loc[test_text[0][0].numpy()][0]

'the'

In [1036]:
word_idx_pd

ID
mdbl       10996
fawc       16260
degussa    12089
woods       8803
hanging    13796
...          ...
rotting    16258
pods       10995
emery       2849
northerly  30979
onomichi   16259

[30979 rows x 1 columns]

In [714]:
np.array(train_weighted_loss[0])

array(None, dtype=object)

In [732]:
kk = np.array(test_accloss.result())
kk

array(2.3583207, dtype=float32)

In [735]:
test_accloss.result().numpy()

2.3583207

In [523]:
# model1.save('./model/emb_layer') #tf.keras.models.load_model
tf.saved_model.save(model1,'./model/emb_layer')
model3 = tf.keras.models.load_model('./model/emb_layer')

TypeError: unsupported callable

In [524]:
model3 = model1.load_weights('./model/emb_layer_weight')

AttributeError: 'CheckpointLoadStatus' object has no attribute 'summary'

In [527]:
model1.emb.set_weights(model3.emb.get_weights())

In [526]:
model3 = model1
model3.load_weights('./model/emb_layer_weight')
model3.summary()

Model: "base_model_1_58"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
masking_58 (Masking)         multiple                  0         
_________________________________________________________________
glove_emb (Embedding)        multiple                  2138112   
_________________________________________________________________


ValueError: You tried to call `count_params` on common_extract, but the layer isn't built. You can build it manually via: `common_extract.build(batch_input_shape)`.

* 在某些參數的情況下0 1是train不起來的但是weight可以
    * alpha = 0.0, beta=0.6
    * init_w = tensorflow.keras.initializers.Constant(value=0.9), init_b = tensorflow.keras.initializers.Constant(value=0.7)
* 0/1 with emb比較容易overfit。如果是weight的比較沒那麼嚴重

* 不用sigmoid或是hard_sigmoid。改良relu + linear，並拆成兩個model，把round前面多加上clip
    * sigmoid中間的變化太快(一瞬間就會掉到0或是1)，改成relu在>0~無限大(linear為了還在0~1)再去clip再round，可以看到每個epoch的變化
* 1st phase的beta一定要>=0.6否則不會動，ones#都會是0
    * beta=1 (放0 1進去): weight init設成1也沒用，但把bias設成1就會一開始都是ones#=1了。weight=1 bias=0.2都匯市0，bias=0.3會是0.87。0.6/0.3都是0。0.8/0.3差不多是0.5(但是train很慢acc進步很慢就是了)。0.5/0.5是從0.01開始往上升 (0 1放進去會比較難train是因為它的變化量太大，一下就是有或沒有，所以clf可能學不好，但如果是weight每次gradient進步的都是一小點就會比較容易上升)
    * beta=0.6 (放0 1進去): 0.8/0.5都是0。0.9/0.8 從0.9一直到0。0.9/0.6差不多是從0.5但又有時候到0.7都是0.0(很難train，Nadam換個opt有時候沒用。EX變成adam 0.9/0.8才有0.96開始但如果0.9/0.75變成0開始。Rmsprop 0.9/0.8又是從0.95開始往下)。但如果都改成傳入weight就都沒問題。ones一開始大概0.5 weight平均，也不會卡住
* 建議: 先訓練embedding weight matrix，但是要看goal是要怎樣的matrix
* 其實他不管幾%都會train得很好，除非固定embedding，或是用更弱的clf
* 若設定alpha，就像是regularizer term (penalty)，設越大drop越多
* 一開始先很多ones，再越來越少個

* 同一個opt若加入transform就會train不起來
* 兩個不同的opt加入transform也會train不起來 (persistent、non-persis都不行)，且與BN無關